---
title: "`r params$set_title`"
chunk_output_type: console
header-includes: \pagenumbering{gobble}
output:
  pdf_document: default
  html_document:
    css: style.css
params:
  set_title: Article figures
---

```{r}

knitr::opts_chunk$set(echo = FALSE,message = FALSE,warning = FALSE,autodep = knitr::dep_prev())

source(here::here('r/prerequisites.R'))

```


***

# Figure 1

```{r eval=T, echo=F}

# MITOMAP gene annotations
mitomap <- read.table('../data/ext/mitomap.txt',header=T,sep='\t',comment.char="",stringsAsFactors=FALSE)

# Separate the MITOMAP file into L- and H- strands
mitomap2 = mitomap[which(mitomap$gene %in% c('TE','ND6','TA','TN','TC','TY','TQ','TS1','TP')),] # L-strand
mitomap1 = mitomap[-which(mitomap$gene %in% c('TE','ND6','TA','TN','TC','TY','TQ','TS1','TP')),] # H-strand

# Import the custom color file
customcolors <- yaml::read_yaml('../data/colors.yaml') %>%
  lapply(function(x) purrr::map_depth(x, purrr::pluck_depth(x)-2, unlist))

allcells_DNA <- read.table('../data/processed_data/combined_coverageDNA.txt',header=T,sep='\t',comment.char="",stringsAsFactors=FALSE)

allcells_RNA <- read.table('../data/processed_data/combined_coverageRNA.txt',header=T,sep='\t',comment.char="",stringsAsFactors=FALSE)

# Import coverage files
lorenz_curve <- read.table('../data/processed_data/lorenz_curve.txt',header=T,sep='\t',stringsAsFactors=FALSE)

coveragebreadth <- read.table('../data/processed_data/coveragebreadth_sample.txt',header=T,sep='\t',stringsAsFactors=FALSE)

# Import oncoprint
oncoprint = read.table('../data/processed_data/oncoprint.txt', row.names = 1, header=F, sep="\t", na.strings="na", stringsAsFactors=FALSE)

# Finds a consecutive positions and concatenates into regions
findIntRuns <- function(run){
  rundiff <- c(1, diff(run))
  difflist <- split(run, cumsum(rundiff>12))
  unlist(lapply(difflist, function(x){
    if(length(x) %in% 1:2) as.character(x) else seq(x[1],x[length(x)])
  }), use.names=FALSE)
}

```

## 1a Schematic of the DLP+

## 1b Circos plot of the median read coverage of the DLP+
```{r panel_1b,fig.height=2.25,fig.width=2.25,message=F,warning=F,eval=T}

s <- which(allcells_RNA$third > allcells_DNA$third)
sreseq <- findIntRuns(s)

ceilingval = max(max(allcells_RNA$third),max(allcells_DNA$third))

# Plot
# Backbone of the circos plot
circos.par(start.degree = 90, gap.degree = 0, track.margin = c(0,0), cell.padding = c(0,0,0,0),"canvas.xlim" = c(-1.12, 1.12), "canvas.ylim" = c(-1.12, 1.12))
circos.initialize(factors = "MT", xlim = c(1, 16569))

# the circos plot track layout for the comparison coverage
circos.track(ylim = c(0, 0.02), track.index = 1, track.height = 0.05)
circos.rect(sreseq, rep(0,length(sreseq)), sreseq, rep(0.02,length(sreseq)), track = 1, col = 'gray18', border = 'gray18')
circos.axis(track = 1, major.at = seq(0, 16569, 1000), labels.cex = 0.8, direction = "outside")

# the circos plot track layout for the median coverage
circos.track(ylim = c(-0.5, ceilingval), track.index = 2, track.height = 0.3)

# The circos plot track of the median coverage for DNA
circos.rect(allcells_DNA$pos, allcells_DNA$first, allcells_DNA$pos, allcells_DNA$third, col = add_transparency("grey", 0.5), border = add_transparency("grey", 0.5), track = 2, lwd = 2)
circos.lines(allcells_DNA$pos, allcells_DNA$median, col = '#4A82B2', track = 2, lwd = 2)

# The circos plot track of the median coverage for RNA
circos.rect(allcells_RNA$pos, allcells_RNA$first, allcells_RNA$pos, allcells_RNA$third, col = "grey", border = "grey", track = 2, lwd = 2)
circos.lines(allcells_RNA$pos, allcells_RNA$median, col = '#F9931D', track = 2, lwd = 2)
circos.yaxis(side = "right", labels.cex = 0.7, at = seq(0, 100, 20), track.index = 2) # break values are automatically calculated

# The circos plot gap in between first
circos.track(ylim = c(0,0.05), track.index = 3, track.height = 0.05)
circos.rect(0, 0, 16569, 0.05, track = 3, col = NULL, border = NULL)

# the circos plot H-strand and labeling the coordinates
circos.track(ylim = c(0,0.04), track.index = 4, track.height = 0.04)
circos.rect(mitomap1[,2], rep(0,31), mitomap1[,3], rep(0.04,31), track = 4, col = mitomap1$col)
circos.rect(mitomap1[-c(1,31),2], rep(0,29), mitomap1[-c(1,31),3], rep(0.04,29), track = 4, col = mitomap1$col[-c(1,40)])

# the circos plot gap in between
circos.track(ylim = c(0,0.02), track.index = 5, track.height = 0.02)
circos.rect(0, rep(0,1), 16569, rep(0.02,1), track = 5, col = "white", border = "white")

# the circos plot L-strand
circos.track(ylim = c(0,0.04), track.index = 6, track.height = 0.04)
circos.rect(mitomap2[,2], rep(0,9), mitomap2[,3], rep(0.04,9), track = 6, col = mitomap2$col, border = mitomap2$col)

# circos.clear()

```


## 1c 
```{r panel_1c,fig.height=1.8,fig.width=2.4,eval=T}

fig1c <- ggplot(lorenz_curve,aes(x=frac_genome,y=tot_coverage,group=interaction(type, sample_id),color=type)) + geom_line(alpha=0.7,linewidth = 0.4) + geom_abline(slope = 1,linetype='dashed') + theme_classic() +
  xlab('Cumulative fraction of genome') + ylab('Cumulative fraction\nof total coverage') + scale_x_continuous(expand = c(0,0)) + scale_y_continuous(expand = c(0,0)) +
  scale_color_manual(breaks = c('DNA','RNA'), values = c('#4984B4','#F59323')) +
  theme(legend.position = c(0.2,0.8),
        axis.text = element_text(size=8),
        axis.title = element_text(size=8),
        legend.key.size = unit(0.5, 'cm'),
        legend.key.height = unit(0.5, 'cm'),
        legend.key.width = unit(0.5, 'cm'),
        legend.title = element_blank())

fig1c

```



## 1d
```{r panel_1d,fig.height=2,fig.width=2.5,eval=T}

fig1d <- ggplot(coveragebreadth,aes(x=index,y=frac_genome,color=type,group=sample_id)) + geom_line() + theme_classic() +
  xlab('Normalized cell index') + ylab('Genome covered (in bp)') + scale_color_manual(values = c('DNA'='#4A83B3','RNA'='#F59322')) +
  scale_x_continuous(expand = c(0,0)) + scale_y_continuous(expand = c(0,0), limits = c(0,16569)) +
  theme(axis.text = element_text(size=8),
        axis.title = element_text(size=8),
        legend.key.size = unit(0.5, 'cm'),
        legend.title = element_blank(),
        legend.position = c(0.85,0.3))

fig1d


```



## 1e
```{r panel_1e,fig.height=2.5,fig.width=2.5,eval=T}

# Sample ID
filtoncoprint = oncoprint[,which(oncoprint[3,] == 'cell_line')]
samplemat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'sample'),])
samplecol = c("#FED9A6","#E7298A","#FB8072","#CAB2D6","#FFFF99","#E78AC3","#CBD5E8","#E31A1C","#F4CAE4",
              "#F0027F","#FDBF6F","#A6CEE3","#E6F5C9","#984EA3")
names(samplecol) = unique(as.character(filtoncoprint[which(rownames(filtoncoprint) == 'sample'),]))

fig1 <- oncoPrint(samplemat,
                  alter_fun = list(
                    background = function(...) NULL,
                    SA039 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA039"], col = NA)),
                    SA1054 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1054"], col = NA)),
                    SA1055 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1055"], col = NA)),
                    SA1056 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1056"], col = NA)),
                    SA1087 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1087"], col = NA)),
                    SA1188 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1188"], col = NA)),
                    SA1292 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1292"], col = NA)),
                    SA906a = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA906a"], col = NA)),
                    SA906b = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA906b"], col = NA)),
                    SA1101a = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1101a"], col = NA)),
                    SA1101b = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1101b"], col = NA)),
                    SA1044 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1044"], col = NA)),
                    SA928 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA928"], col = NA)),
                    SA1090 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1090"], col = NA))
                  ), col = samplecol, top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = 'Sample', show_heatmap_legend = FALSE)

lgd1 = Legend(labels = names(samplecol), legend_gp = gpar(fill = samplecol), labels_gp = gpar(fontsize = 4), title = "", direction = "horizontal")

# Sample type
datamat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'Datatype'),])
datacol = c(cell_line = "deepskyblue3")
fig2 <- oncoPrint(datamat,
                  alter_fun = list(
                    background = function(...) NULL,
                    cell_line = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                               gp = gpar(fill = datacol["cell_line"], col = NA))
                  ), col = rep('grey',3), top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = 'Sample type', show_heatmap_legend = FALSE)

newnames_datacol = names(datacol)
newnames_datacol[which(newnames_datacol == 'cell_line')] = 'Cell Line'
lgd2 = Legend(labels = newnames_datacol, legend_gp = gpar(fill = datacol), labels_gp = gpar(fontsize = 6), title = "", direction = "horizontal")

# Tissue type
tissuemat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'Tissuetype'),])
tissuecol = c(Breast = "#000296", Lymphoblastoid = "#FBCE78", Cervical = "#F4D8DD", Ovary = "khaki4")
fig3 <- oncoPrint(tissuemat,
                  alter_fun = list(
                    background = function(...) NULL,
                    Breast = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                            gp = gpar(fill = tissuecol["Breast"], col = NA)),
                    Lymphoblastoid = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                            gp = gpar(fill = tissuecol["Lymphoblastoid"], col = NA)),
                    Cervical = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                                    gp = gpar(fill = tissuecol["Cervical"], col = NA)),
                    Ovary = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                           gp = gpar(fill = tissuecol["Ovary"], col = NA))
                  ), col = tissuecol, top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = 'Tissue type', show_heatmap_legend = FALSE)

lgd3 = Legend(labels = names(tissuecol), legend_gp = gpar(fill = tissuecol), labels_gp = gpar(fontsize = 6), title = "", direction = "horizontal")

# Matching scRNA-seq
scrnamat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'scRNA.seq'),])
presencecol = c(Yes = "black", No = "ivory")
fig4 <- oncoPrint(scrnamat,
                  alter_fun = list(
                    background = function(...) NULL,
                    Yes = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                         gp = gpar(fill = presencecol["Yes"], col = NA)),
                    No = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                        gp = gpar(fill = presencecol["No"], col = NA))
                  ), col = presencecol, top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = 'scRNA-seq', show_heatmap_legend = FALSE)

lgd4 = Legend(labels = names(presencecol), legend_gp = gpar(fill = presencecol), labels_gp = gpar(fontsize = 6), title = "", direction = "horizontal")

# Histogram
ncellmat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'Ncells'),])
ncellmat = apply(ncellmat, 2, as.numeric)
ncellmat = t(ncellmat)
ha = HeatmapAnnotation(ncell = anno_barplot(ncellmat, beside = TRUE, gp = gpar(fill = 1), height = unit(4, "cm")), show_annotation_name = FALSE)
plot(ha)

ht_list = fig1 %v% fig2 %v% fig3 %v% fig4
print(ht_list)

pd1 <- packLegend(lgd1, direction = "vertical")
grid::grid.newpage()
draw(pd1)

# Sample type
datacol = c('Cell Line' = "deepskyblue3", 'Tumor (PDX)' = "olivedrab3", 'Tumor (Primary)' = "pink3")
lgd2 = Legend(labels = datacol, legend_gp = gpar(fill = datacol), labels_gp = gpar(fontsize = 6), title = "", direction = "horizontal")

pd2 <- packLegend(lgd2, lgd3, lgd4, direction = 'horizontal')
grid::grid.newpage()
draw(pd2)

## Tumors
filtoncoprint = oncoprint[,which((oncoprint[3,] != 'cell_line') & (as.character(oncoprint[5,]) != 'Treated'))]

# Sample ID
samplemat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'sample'),])
samplecol = c("#E3B172","#CCDCDB","#76BFBA","#00427D","#0186BA","#9404D2","#998884","#DED3CC","#55B700","#BB937E","#915549","#D38842","#8B3415","#362F37","#EEFFAA")
names(samplecol) = unique(as.character(filtoncoprint[which(rownames(filtoncoprint) == 'sample'),]))

fig1 <- oncoPrint(samplemat,
                  alter_fun = list(
                    background = function(...) NULL,
                    SA1035 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1035"], col = NA)),
                    SA1052 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1052"], col = NA)),
                    SA1142 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1142"], col = NA)),
                    SA501 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA501"], col = NA)),
                    SA530 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA530"], col = NA)),
                    SA535 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA535"], col = NA)),
                    SA604 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA604"], col = NA)),
                    SA605 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA605"], col = NA)),
                    SA609 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA609"], col = NA)),
                    SA610 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA610"], col = NA)),
                    OV022 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["OV022"], col = NA)),
                    OV081 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["OV081"], col = NA)),
                    SA1047 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1047"], col = NA)),
                    SA1135 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1135"], col = NA)),
                    SA1162 = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = samplecol["SA1162"], col = NA))
                  ), col = samplecol, top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = '', show_heatmap_legend = FALSE)

lgd1 = Legend(labels = gsub('_','-',names(samplecol)), legend_gp = gpar(fill = samplecol), labels_gp = gpar(fontsize = 4), title = "", direction = "horizontal")

# Sample type
datamat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'Datatype'),])
datacol = c(PDX = "olivedrab3", primary = "pink3")
fig2 <- oncoPrint(datamat,
                  alter_fun = list(
                    background = function(...) NULL,
                    PDX = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                         gp = gpar(fill = datacol["PDX"], col = NA)),
                    primary = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                             gp = gpar(fill = datacol["primary"], col = NA))
                  ), col = rep('grey',3), top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = '', show_heatmap_legend = FALSE)

newnames_datacol = names(datacol)
newnames_datacol[which(newnames_datacol == 'PDX')] = 'Tumor (PDX)'
newnames_datacol[which(newnames_datacol == 'primary')] = 'Tumor (Primary)'
lgd2 = Legend(labels = newnames_datacol, legend_gp = gpar(fill = datacol), labels_gp = gpar(fontsize = 4), title = "", direction = "horizontal")

# Tissue type
tissuemat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'Tissuetype'),])
tissuecol = c(Breast = "#000296", Ovary = "khaki4")
fig3 <- oncoPrint(tissuemat,
                  alter_fun = list(
                    background = function(...) NULL,
                    Breast = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                            gp = gpar(fill = tissuecol["Breast"], col = NA)),
                    Ovary = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                           gp = gpar(fill = tissuecol["Ovary"], col = NA))
                  ), col = tissuecol, top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = '', show_heatmap_legend = FALSE)

lgd3 = Legend(labels = names(tissuecol), legend_gp = gpar(fill = tissuecol), labels_gp = gpar(fontsize = 4), title = "", direction = "horizontal")

# Treatment status
treatmentmat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'Cisplatin'),])
treatmentcol= c(Untreated = "violet", Treated = "green4")
fig4 <- oncoPrint(treatmentmat,
                  alter_fun = list(
                    background = function(...) NULL,
                    Untreated = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                               gp = gpar(fill = treatmentcol["Untreated"], col = NA)),
                    Treated = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                             gp = gpar(fill = treatmentcol["Treated"], col = NA))
                  ), col = treatmentcol, top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", show_heatmap_legend = FALSE)


lgd4 = Legend(labels = names(treatmentcol), legend_gp = gpar(fill = treatmentcol), labels_gp = gpar(fontsize = 4), title = "", direction = "horizontal")


# Matching scRNA-seq
scrnamat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'scRNA.seq'),])
presencecol = c(Yes = "black", No = "ivory")
fig4 <- oncoPrint(scrnamat,
                  alter_fun = list(
                    background = function(...) NULL,
                    Yes = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                         gp = gpar(fill = presencecol["Yes"], col = NA)),
                    No = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                        gp = gpar(fill = presencecol["No"], col = NA))
                  ), col = presencecol, top_annotation = NULL, right_annotation = NULL, show_pct = FALSE, row_names_side = "left", row_labels = '', show_heatmap_legend = FALSE)

lgd4 = Legend(labels = names(presencecol), legend_gp = gpar(fill = presencecol), labels_gp = gpar(fontsize = 4), title = "", direction = "horizontal")

# Histogram
ncellmat = as.matrix(filtoncoprint[which(rownames(filtoncoprint) == 'Ncells'),])
ncellmat = apply(ncellmat, 2, as.numeric)
ncellmat = t(ncellmat)
ha = HeatmapAnnotation(ncell = anno_barplot(ncellmat, beside = TRUE, gp = gpar(fill = 1), height = unit(4, "cm"), axis = F), show_annotation_name = FALSE)
plot(ha)

ht_list = fig1 %v% fig2 %v% fig3 %v% fig4
print(ht_list)

pd1 <- packLegend(lgd1, direction = "vertical")
pd2 <- packLegend(lgd2, lgd3, lgd4, direction = 'horizontal')

grid::grid.newpage()
draw(pd1)
grid::grid.newpage()
draw(pd2)

```


***

# Figure 2

```{r eval=T, echo=F}

# Import the master result file across all cells
resultfile = read.csv(paste0('../data/processed_data/final_master.tsv'), header=T, sep='\t', stringsAsFactors=F)

# Subset the cell image data
micron_resultfile = resultfile[which(resultfile$spotter == "rachael"),] # Sort by rachael spotter
pixel_resultfile = resultfile[which(resultfile$spotter == "deckard"),] # Sort by deckard spotter

resultfile = resultfile[-which((resultfile$library_label == 'SA928-ALL') | (resultfile$group %in% c("treated TNBC PDX(SA1035)","treated TNBC PDX(SA535)","treated TNBC PDX(SA609)"))),]
resultfile$library_label = gsub('OV_022','OV022',resultfile$library_label)
resultfile$library_label = gsub('OV_081','OV081',resultfile$library_label)

# Annotating the timepoints
SA039_passages = list("A73044B"="p25","A96199B"="p30","A73047D"="p51","A96226B"="p60") # SA039
SA906a_passages = list("A96180B"="p10","A96210C"="p15","A96216A"="p25","A96175C"="p30","A96183C"="p40","A96186A"="p50","A96146A"="p57") # SA906a
SA906b_passages = list("A96215A"="p20","A96211C"="p30","A96172B"="p35","A96155B"="p40","A96220B"="p45","A96228B"="p50","A96149B"="p55") # SA906b
SA1101a_passages = list("A96225B"="p10","A96162A"="p15","A96155C"="p20","A96240A"="p25","A96225C"="p30","A96193B"="p35","A96183B"="p40","A96240B"="p45","A96181C"="p50") # SA1101a
SA1101b_passages = list("A96233B"="p15","A96190A"="p20","A95634B"="p25","A95668B"="p30","A95663B"="p35","A96118A"="p40","A96120A"="p41","A96114A"="p45","A96124B"="p45","A95697B"="p50","A98271A"="p50","A95700A"="p55","A95723A"="p55","A98295B"="p60","A98274B"="p60") # SA1101b
SA1035_passages = list('A95626A'='p4','A95646A'='p4','A95731B'='p5','A95623B'='p6','A98290B'='p7','A98251B'='p7','A98253A'='p8','A98256A'='p8')
SA609_passages = list('A95724A'='p3','A96201A'='p3','A95632D'='p6','A96219B'='p10')
SA501_passages = list('A95621B'='p2','A96109A'='p2','A96171A'='p2','A95670A'='p5','A96213A'='p6','A95670B'='p6','A96174A'='p11','A96187A'='p11','A96173A'='p15')

cell_line_passages = c(SA039_passages,SA906a_passages,SA906b_passages,SA1101a_passages,SA1101b_passages)
tumor_passages = c(SA1035_passages,SA609_passages,SA501_passages)

# Adding timepoint information
last = apply(data.frame(resultfile$library_id),1,function(x) strsplit(x,'-')[[1]][2])

# Filtering for timepoint file in cell lines
CL_timepointfile = resultfile
CL_timepointfile$timepoint = as.character(cell_line_passages[last])
CL_timepointfile = CL_timepointfile[-which(CL_timepointfile$timepoint == 'NULL'),]
CL_timepointfile$timepoint = factor(CL_timepointfile$timepoint, levels = paste0("p",sort(as.numeric(unique(gsub("p","",CL_timepointfile$timepoint))))))

# Filtering for timepoint file in tumors
T_timepointfile = resultfile
T_timepointfile$timepoint = as.character(tumor_passages[last])
T_timepointfile = T_timepointfile[-which(T_timepointfile$timepoint == 'NULL'),]
T_timepointfile = T_timepointfile[-which(T_timepointfile$group == "HGSC cell line(SA1090)"),]
T_timepointfile$timepoint = factor(T_timepointfile$timepoint, levels = paste0("p",sort(as.numeric(unique(gsub("p","",T_timepointfile$timepoint))))))

# Import SA906a TP53-/- 184-hTERT epithelial cell line master file
SA906a_master <- read.table('../data/master_files/SA906a-ALL_master.tsv',header=T,sep='\t',quote='',stringsAsFactors=F)

# Import SA039 184-hTERT epithelial cell line passage data
passage_file <- read.table('../data/processed_data/SA039_passage.tsv',header=T,sep='\t',quote='',stringsAsFactors=F)

# Import oncoprint
oncoprint = read.table('../data/processed_data/oncoprint.txt', row.names = 1, header=F, sep="\t", na.strings="na", stringsAsFactors=FALSE)

```


## 2a

```{r panel_2a,fig.height=3,fig.width=5,eval=T}
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Serial passage experiment of 184-hTERT breast epithelial cell line
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

PDXfile = resultfile[grep('PDX',resultfile$group),]
primaryfile = resultfile[c(grep('primary',resultfile$group),grep('SPECTRUM',resultfile$group)),]
cell_linefile = resultfile[-c(grep('primary',resultfile$group),grep('PDX',resultfile$group),grep('SPECTRUM',resultfile$group)),]

## Distribution of per-cell mtDNA copy number with perturbations
oncoprint = oncoprint[,-which(as.character(oncoprint[5,]) == 'Treated')]

# Library ID and tissue type
metadat = data.frame(paste0(oncoprint[which(rownames(oncoprint) == 'sample'),],'-',oncoprint[which(rownames(oncoprint) == 'library_id'),]),as.character(oncoprint[which(rownames(oncoprint) == 'Tissuetype'),]))
colnames(metadat) = c('library_id','tissue')

## Calculate the median mtDNA copy number per library
ordergroup = resultfile %>% group_by(library_label) %>% mutate(medcn=median(copynumber)) %>% select(library_label, medcn) %>% distinct()
ordergroup = ordergroup[order(ordergroup$medcn, decreasing = T),]

finalorder = ordergroup$library_label

medline = data.frame(t(c(1,1,1,1)))
colnames(medline) = c("cell_id","center","median","library_label")
for (eachlib in finalorder){
  currpoints = resultfile[which(resultfile$library_label == eachlib),]
  if (nrow(currpoints) < 10){
    tempres2 = data.frame(currpoints$cell_id[order(currpoints$copynumber)])
  }else if(nrow(currpoints) < 30){
    tempres2 = data.frame(currpoints$cell_id[order(currpoints$copynumber)[(as.integer(length(currpoints$copynumber)/2)-as.integer(length(currpoints$copynumber)*0.3)):(as.integer(length(currpoints$copynumber)/2)+as.integer(length(currpoints$copynumber)*0.3))]])
  }else{
    tempres2 = data.frame(currpoints$cell_id[order(currpoints$copynumber)[(as.integer(length(currpoints$copynumber)/2)-as.integer(length(currpoints$copynumber)*0.2)):(as.integer(length(currpoints$copynumber)/2)+as.integer(length(currpoints$copynumber)*0.2))]])
  }
  tempres2$center = currpoints$cell_id[order(currpoints$copynumber)[as.integer(length(currpoints$copynumber)/2)]]
  tempres2$median = median(currpoints$copynumber[which(currpoints$library_label == eachlib)])
  tempres2$library_label = eachlib
  colnames(tempres2) = colnames(medline)
  medline = rbind(medline, tempres2)
}
medline = medline[-1,]

medline$library_label = factor(medline$library_label, levels=finalorder)

for (eachtype in c('PDX','primary','cell_line')){
  print(eachtype)
  currfile = get(paste0(eachtype,'file'))
  currorder = finalorder[which(finalorder %in% unique(currfile$library_label))]
  currmedline = medline[which(medline$library_label %in% unique(currfile$library_label)),]
  currmedline$library_label = factor(currmedline$library_label,currorder)
  
  res = currfile %>% group_by(library_label) %>% mutate(medcn=median(copynumber))
  res$library_label = factor(res$library_label, levels=currorder)
  res$tissue = metadat$tissue[match(res$library_label,metadat$library_id)]
  res$tissue = factor(res$tissue, levels=unique(res$tissue))
  
  if (eachtype == 'cell_line'){
    plotcell_line = ggplot() +
      geom_rect(data = subset(res,library_label %in% currorder[seq(1,length(currorder),by=2)]),fill = "lightgrey",xmin = -Inf,xmax = Inf,ymin = -Inf,ymax = Inf,alpha=0.1,inherit.aes = F) +
      geom_point(data = res, aes(x = reorder(cell_id,copynumber), y = copynumber, color=tissue), size = 0.6, alpha=0.3, inherit.aes = F) +
      scale_color_manual(values=customcolors$tissuetypes) +
      theme_classic() +
      geom_point(data=currmedline, inherit.aes = F, aes(x=cell_id,y=median), size = 0.4, shape = 15, col="black") +
      facet_wrap(library_label~.,nrow=1, scales = "free_x") +
      coord_cartesian(clip = "off") +
      ylab("Per-cell\nmtDNA copy number") +
      scale_y_continuous(limits = c(1,7500)) +
      scale_x_discrete(breaks=unique(currmedline$center), labels=currorder) +
      theme(axis.text.x = element_text(size=5, angle = 55, hjust = 1, vjust = 1),
            axis.text.y = element_text(size=8),
            axis.title.x = element_blank(),
            axis.title.y = element_text(size=10),
            strip.background = element_blank(),
            strip.text.x = element_blank(),
            panel.grid.minor = element_blank(),
            panel.grid.major.x = element_blank(),
            panel.spacing = unit(0.1, "lines"),
            legend.position = "none",
            plot.margin = margin(l = 5, r = 5),
            rect = element_rect(fill = "transparent",colour = NA),
            plot.background = element_rect(fill = "transparent",colour = NA)
      )
  }else if (eachtype == 'PDX'){
    plotPDX = ggplot() +
      geom_rect(data = subset(res,library_label %in% currorder[seq(1,length(currorder),by=2)]),fill = "lightgrey",xmin = -Inf,xmax = Inf,ymin = -Inf,ymax = Inf,alpha=0.1,inherit.aes = F) +
      geom_point(data = res, aes(x = reorder(cell_id,copynumber), y = copynumber, color=tissue), size = 0.6, alpha=0.3, inherit.aes = F) +
      scale_color_manual(values=customcolors$tissuetypes) +
      theme_classic() +
      geom_point(data=currmedline, inherit.aes = F, aes(x=cell_id,y=median), size = 0.4, shape = 15, col="black") +
      facet_wrap(library_label~.,nrow=1, scales = "free_x") +
      coord_cartesian(clip = "off") +
      ylab("Per-cell mtDNA copy number") +
      scale_y_continuous(limits = c(1,7500)) +
      scale_x_discrete(breaks=unique(currmedline$center), labels=currorder) +
      theme(axis.text.x = element_text(size=5, angle = 55, hjust = 1, vjust = 1),
            axis.text.y = element_blank(),
            axis.title = element_blank(),
            strip.background = element_blank(),
            strip.text.x = element_blank(),
            panel.grid.minor = element_blank(),
            panel.grid.major.x = element_blank(),
            panel.spacing = unit(0.1, "lines"),
            legend.position = "none",
            plot.margin = margin(l = 16, r = 5),
            rect = element_rect(fill = "transparent",colour = NA),
            plot.background = element_rect(fill = "transparent",colour = NA)
      )
  }else if (eachtype == 'primary'){
    plotprimary = ggplot() +
      geom_rect(data = subset(res,library_label %in% currorder[seq(1,length(currorder),by=2)]),fill = "lightgrey",xmin = -Inf,xmax = Inf,ymin = -Inf,ymax = Inf,alpha=0.1,inherit.aes = F) +
      geom_point(data = res, aes(x = reorder(cell_id,copynumber), y = copynumber, color=tissue), size = 0.6, alpha=0.3, inherit.aes = F) +
      scale_color_manual(values=customcolors$tissuetypes) +
      theme_classic() +
      geom_point(data=currmedline, inherit.aes = F, aes(x=cell_id,y=median), size = 0.4, shape = 15, col="black") +
      facet_wrap(library_label~.,nrow=1, scales = "free_x") +
      coord_cartesian(clip = "off") +
      scale_y_continuous(limits = c(1,7500)) +
      scale_x_discrete(breaks=unique(currmedline$center), labels=currorder) +
      theme(axis.text.x = element_text(size=5, angle = 55, hjust = 1, vjust = 1),
            axis.text.y = element_blank(),
            axis.title = element_blank(),
            strip.background = element_blank(),
            strip.text.x = element_blank(),
            panel.grid.minor = element_blank(),
            panel.grid.major.x = element_blank(),
            panel.spacing = unit(0.1, "lines"),
            legend.position = "none",
            plot.margin = margin(l = 20, r = 5),
            rect = element_rect(fill = "transparent",colour = NA),
            plot.background = element_rect(fill = "transparent",colour = NA)
      )
  }
}

plotcell_line
plotPDX
plotprimary

# Legend
legend = Legend(labels = names(customcolors$tissuetypes), legend_gp = gpar(fill = as.character(customcolors$tissuetypes)), labels_gp = gpar(fontsize = 6), title = "", direction = "vertical", nrow = 1)
grid::grid.newpage()
draw(legend)


```



## 2b

```{r panel_2b,fig.height=2,fig.width=4,eval=T}

pvals = c()
for (eachsample in unique(CL_timepointfile$group)){
  filtresultfile = CL_timepointfile[which(CL_timepointfile$group == eachsample),]
  pval = kruskal.test(copynumber ~ timepoint, data = filtresultfile)$p.value
  pvals = append(pvals, pval)
}

summarydata = CL_timepointfile %>% group_by(group,timepoint) %>% mutate(medcopynumber = median(copynumber), sdcopynumber = (sd(copynumber)), secopynumber = (sdcopynumber/sqrt(n())), n = n()) %>% select(group,timepoint,medcopynumber,sdcopynumber,secopynumber,n) %>% distinct()
colnames(summarydata) = c("group","timepoint","median_copynumber","sd_copynumber","se_copynumber","n")

# Fixing the group labels
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA1101a)')] = 'TP53-/-(SA1101a)'
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA1101b)')] = 'TP53-/-(SA1101b)'
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA906a)')] = 'TP53-/-(SA906a)'
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA906b)')] = 'TP53-/-(SA906b)'

fig2b <- ggplot(subset(summarydata,group == '184hTERT'),aes(x=timepoint,y=median_copynumber)) + geom_errorbar(aes(ymin=median_copynumber-se_copynumber, ymax=median_copynumber+se_copynumber), width=.2, position=position_dodge(.9)) +
  geom_line(aes(group=group,col=group),size=0.6) + geom_point(size=0.6) + theme_classic() +
  xlab("Timepoints") + ylab(label = "mtDNA copy number") +
  scale_y_continuous(limits = c(0,2000), expand=c(0,0)) +
  theme(legend.position = 'none',
        axis.title = element_text(size=9),
        axis.text = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=90, vjust = 0.5, hjust = 0.5, size=6))

fig2b


```


## 2c

```{r panel_2c,fig.height=2,fig.width=4,eval=T}

summarydata = CL_timepointfile %>% group_by(group,timepoint) %>% mutate(medcopynumber = median(copynumber), sdcopynumber = (sd(copynumber)), secopynumber = (sdcopynumber/sqrt(n())), n = n()) %>% select(group,timepoint,medcopynumber,sdcopynumber,secopynumber,n) %>% distinct()
colnames(summarydata) = c("group","timepoint","median_copynumber","sd_copynumber","se_copynumber","n")

# Fixing the group labels
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA1101a)')] = 'TP53-/-(SA1101a)'
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA1101b)')] = 'TP53-/-(SA1101b)'
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA906a)')] = 'TP53-/-(SA906a)'
summarydata$group[which(summarydata$group == '184hTERT(TP53-/-)(SA906b)')] = 'TP53-/-(SA906b)'

fig2c <- ggplot(subset(summarydata,group != '184hTERT'),aes(x=timepoint,y=median_copynumber)) + geom_errorbar(aes(ymin=median_copynumber-se_copynumber, ymax=median_copynumber+se_copynumber), width=.2, position=position_dodge(.9)) +
  geom_line(aes(group=group,col=group),size=0.6) + geom_point(size=0.6) + theme_classic() +
  xlab("Timepoints") + ylab(label = "mtDNA copy number") +
  scale_y_continuous(limits = c(0,2500), expand=c(0,0)) +
  theme(axis.title = element_text(size=9),
        axis.text = element_text(size=6),
        axis.text.x = element_text(angle=90, vjust = 0.5, hjust = 0.5, size=6),
        axis.title.x = element_blank())

fig2c_legend <- as_ggplot(get_legend(fig2c))

fig2c <- fig2c + theme(legend.position = 'none')
fig2c
fig2c_legend

```


## 2d

```{r panel_2d,fig.height=2,fig.width=4,eval=T}

pvals = c()
for (eachsample in unique(T_timepointfile$group)){
  filtresultfile = T_timepointfile[which(T_timepointfile$group == eachsample),]
  pval = kruskal.test(copynumber ~ timepoint, data = filtresultfile)$p.value
  pvals = append(pvals, pval)
}

summarydata = T_timepointfile %>% group_by(group,timepoint) %>% mutate(medcopynumber = median(copynumber), sdcopynumber = (sd(copynumber)), secopynumber = (sdcopynumber/sqrt(n())), n = n()) %>% select(group,timepoint,medcopynumber,sdcopynumber,secopynumber,n) %>% distinct()
colnames(summarydata) = c("group","timepoint","median_copynumber","sd_copynumber","se_copynumber","n")

# Fixing the group labels
summarydata$group[which(summarydata$group == 'untreated TNBC PDX(SA1035)')] = 'TNBC PDX(SA1035)'
summarydata$group[which(summarydata$group == 'untreated TNBC PDX(SA535)')] = 'TNBC PDX(SA535)'
summarydata$group[which(summarydata$group == 'untreated TNBC PDX(SA609)')] = 'TNBC PDX(SA609)'

fig2d <- ggplot(summarydata,aes(x=timepoint,y=median_copynumber)) + geom_errorbar(aes(ymin=median_copynumber-se_copynumber, ymax=median_copynumber+se_copynumber), width=.2, position=position_dodge(.9)) +
  geom_line(aes(group=group,col=group),size=0.6) + geom_point(size=0.6) + theme_classic() +
  xlab("Timepoints") + ylab(label = "mtDNA copy number") +
  scale_y_continuous(limits = c(0,1800), expand=c(0,0)) +
  theme(axis.title = element_text(size=9),
        axis.text = element_text(size=6),
        axis.text.x = element_text(angle=90, vjust = 0.5, hjust = 0.5, size=6),
        axis.title.x = element_blank())

fig2d_legend <- as_ggplot(get_legend(fig2d))

fig2d <- fig2d + theme(legend.position = 'none')
fig2d
fig2d_legend


```


## 2e

```{r panel_2e,fig.height=2,fig.width=2,eval=T}

filtmicron_resultfile = micron_resultfile[which(micron_resultfile$library_id == 'SA906-A96155B'),]

fig2e <- ggplot(subset(filtmicron_resultfile, ploidylab == 2), aes(x=diameter, y=copynumber)) + stat_density2d(aes(fill = ..density..^0.25), geom = "tile", contour = FALSE, n = 400) +
  geom_point(size=0.8, alpha = 0.3, shape = 16) + stat_cor(size = 4) + scale_fill_continuous(low = "white", high = "dodgerblue4") + geom_smooth(method = lm, size = 0.5, col= 'red', linetype = "dashed") + theme_classic() +
  xlab('Cell diameter (in microns)') + ylab('mtDNA copy number') + ggtitle('SA906-A96155B') +
  theme(axis.text = element_text(size=10),
        axis.title = element_text(size=10),
        title = element_text(size=8),
        legend.position = 'none')

fig2e


```


## 2f

```{r panel_2f,fig.height=2,fig.width=2,eval=T}

filtmicron_resultfile = micron_resultfile[which(micron_resultfile$library_id == 'SA1035X5XB03015-A95623A'),]

fig2f <- ggplot(subset(filtmicron_resultfile, ploidylab == 2), aes(x=diameter, y=copynumber)) + stat_density2d(aes(fill = ..density..^0.25), geom = "tile", contour = FALSE, n = 400) +
  geom_point(size=0.8, alpha = 0.3, shape = 16) + stat_cor(size = 4) + scale_fill_continuous(low = "white", high = "dodgerblue4") + geom_smooth(method = lm, size = 0.5, col= 'red', linetype = "dashed") + theme_classic() +
  xlab('Cell diameter (in microns)') + ylab('mtDNA copy number') + ggtitle('SA1035-A95623A') +
  theme(axis.text = element_text(size=10),
        axis.title = element_text(size=10),
        title = element_text(size=8),
        legend.position = 'none')

fig2f


```


## 2g,h

```{r panel_2gh,fig.height=3,fig.width=2,eval=T}

colorlist = c('184hTERT(TP53-/-)(SA906a)'='#A6CEE3',
              'T-47D cell line(SA1044)'='#E7298A',
              '184hTERT(BRCA2-/-;TP53-/-)(SA1056)'='#FFFF99',
              'HGSC cell line(SA1090)'='#CBD5E8',
              '184hTERT'='#FED9A6',
              '184hTERT(TP53-/-)(SA906b)'='#E6F5C9',
              '184hTERT(TP53-/-)(SA1101a)'='#E31A1C',
              '184hTERT(TP53-/-)(SA1101b)'='#F4CAE4',
              '184hTERT(BRCA1-/-;TP53-/-)(SA1054)'='#FB8072',
              '184hTERT(BRCA2+/-;TP53-/-)(SA1188)'='#F0027F',
              'GM18507'='#984EA3',
              'TNBC PDX(SA501)'='#00427D',
              'TNBC PDX(SA1142)'='#76BFBA',
              'untreated TNBC PDX(SA609)'='#55B700',
              'untreated TNBC PDX(SA535)'='#9404D2',
              'untreated TNBC PDX(SA1035)'='#E3B172',
              'HGSC primary(SA1047)'='#8B3415',
              'HGSC primary(SA1162)'='#EEFFAA',
              'HGSC primary(OV-022)'='#915549',
              'HGSC primary(OV-081)'='#D38842',
              'TNBC primary(SA1135)'='#362F37',
              'white'='white')

#### cell size ~ mtDNA ploidy (without lymphoblastoid and 184hTERT)
micron_resultfile = resultfile[which(resultfile$spotter == "rachael"),] # Sort by rachael spotter
pixel_resultfile = resultfile[which(resultfile$spotter == "deckard"),] # Sort by deckard spotter

totcells = c()

# Actual values
siggroups = data.frame(t(c(1,1,1,1,1,1,1,1,1)))
colnames(siggroups) = c("library_id","library_label","numbcell","group","coefficient","lowerconfint","upperconfint","spotter","pval")

for (eachlib in unique(micron_resultfile$library_id)){
  filtmicron_result <- micron_resultfile[which((micron_resultfile$library_id == eachlib) & (micron_resultfile$ploidylab == 2)),]
  if (nrow(filtmicron_result) >= 12){
    totcells = append(totcells, nrow(filtmicron_result))
    testres = cor.test(filtmicron_result$diameter, filtmicron_result$copynumber, method = "pearson")
    tempres = data.frame(eachlib,unique(filtmicron_result$library_label),nrow(filtmicron_result),unique(micron_resultfile$group[which(micron_resultfile$library_id == eachlib)]),testres$estimate,testres$conf.int[1],testres$conf.int[2],unique(filtmicron_result$spotter),testres$p.value)
    colnames(tempres) = colnames(siggroups)
    siggroups = rbind(siggroups, tempres)
  }
}

for (eachlib in unique(pixel_resultfile$library_id)){
  filtpixel_result <- pixel_resultfile[which((pixel_resultfile$library_id == eachlib) & (pixel_resultfile$ploidylab == 2)),]
  if (nrow(filtpixel_result) >= 12){
    totcells = append(totcells, nrow(filtpixel_result))
    testres = cor.test(filtpixel_result$diameter, filtpixel_result$copynumber, method = "pearson")
    tempres = data.frame(eachlib,unique(filtpixel_result$library_label),nrow(filtpixel_result),unique(pixel_resultfile$group[which(pixel_resultfile$library_id == eachlib)]),testres$estimate,testres$conf.int[1],testres$conf.int[2],unique(filtpixel_result$spotter),testres$p.value)
    colnames(tempres) = colnames(siggroups)
    siggroups = rbind(siggroups, tempres)
  }
}
siggroups = siggroups[-1,]

deckarddups = siggroups[which((siggroups$library_id %in% c("SA906-A96172B","SA1101-A95663B","SA1142X2XB09021-A95700B","SA609X10XB02454-A96219B") & (siggroups$spotter == 'deckard'))),]
siggroups = siggroups[-which((siggroups$library_id %in% c("SA906-A96172B","SA1101-A95663B","SA1142X2XB09021-A95700B","SA609X10XB02454-A96219B") & (siggroups$spotter == 'deckard'))),]
siggroups$padj <- p.adjust(siggroups$pval,method='BH')
siggroups$signif = siggroups$padj <= 0.05
siggroups$shortlibid = siggroups$library_id
siggroups$shortlibid[which(apply(data.frame(siggroups$shortlibid),1,function(x) length(strsplit(x,'-')[[1]])) == 2)] = apply(data.frame(siggroups$shortlibid),1,function(x)strsplit(x,'-')[[1]][2])[which(apply(data.frame(siggroups$shortlibid),1,function(x) length(strsplit(x,'-')[[1]])) == 2)]

ordering = siggroups %>% group_by(group) %>% mutate(medcoeff = median(coefficient)) %>% select(group, medcoeff) %>% distinct()

reorderedgroups = data.frame(t(rep(1,length(colnames(siggroups)))))
colnames(reorderedgroups) = colnames(siggroups)
for (eachgroup in ordering$group[order(ordering$medcoeff, decreasing = F)]){
  currgroup = siggroups[which(siggroups$group == eachgroup),]
  currgroup = currgroup[order(currgroup$coefficient, decreasing = F),]
  reorderedgroups = rbind(reorderedgroups, currgroup)
}
reorderedgroups = reorderedgroups[-1,]

reorderedgroups$library_id[grep('SPECTRUM',reorderedgroups$group)] = paste0(gsub('SPECTRUM-','',reorderedgroups$group[grep('SPECTRUM',reorderedgroups$group)]),'-',reorderedgroups$library_id[grep('SPECTRUM',reorderedgroups$group)])
reorderedgroups$library_id = factor(reorderedgroups$library_id, level = unique(reorderedgroups$library_id))
labelcolor = data.frame(ordering$group[order(ordering$medcoeff, decreasing = F)],colorlist[match(ordering$group[order(ordering$medcoeff, decreasing = F)],names(colorlist))])
colnames(labelcolor) = c('group','colors')

PDXfile = reorderedgroups[grep('PDX',reorderedgroups$group),]
primaryfile = reorderedgroups[c(grep('primary',reorderedgroups$group),grep('SPECTRUM',reorderedgroups$group)),]
cell_linefile = reorderedgroups[-c(grep('primary',reorderedgroups$group),grep('SPECTRUM',reorderedgroups$group),grep('PDX',reorderedgroups$group)),]
cell_linefile$fillcol = 'white'
cell_linefile$fillcol[which(cell_linefile$signif == 1)] = cell_linefile$group[which(cell_linefile$signif == 1)]
cell_linefile$library_id = factor(cell_linefile$library_id,levels = unique(cell_linefile$library_id))

figcell_line = ggplot() +
  geom_pointrange(data=cell_linefile, aes(x=library_id, y=coefficient, ymin=lowerconfint, ymax=upperconfint, color=group),size=0.2) +
  geom_errorbar(data=cell_linefile, aes(x=library_id, y=coefficient, ymin=lowerconfint, ymax=upperconfint, color=group),width=0.4) +
  geom_point(data=cell_linefile, aes(x=library_id, y=coefficient, color=fillcol), size=0.2) +
  ylab("Pearson correlation coefficient") +
  geom_hline(yintercept = 0, lty='dashed') +
  scale_y_continuous(limits = c(-0.5,1.1)) +
  theme_bw() +
  coord_flip() +
  scale_x_discrete(limits = levels(cell_linefile$library_id), labels = cell_linefile$shortlibid) +
  scale_color_manual(values=colorlist) +
  theme(axis.title.y = element_blank(),
        axis.text = element_text(size=7),
        axis.title.x = element_text(size=8),
        strip.text.x = element_text(size=7),
        strip.background = element_blank(),
        axis.line = element_line(size = 0.5, colour = "black"),
        panel.grid.minor = element_blank(),
        legend.position = "none",
        panel.border = element_blank())

PDXfile$fillcol = 'white'
PDXfile$fillcol[which(PDXfile$signif == 1)] = PDXfile$group[which(PDXfile$signif == 1)]
PDXfile$library_label = factor(PDXfile$library_label,levels = unique(PDXfile$library_label))

figPDX = ggplot() +
  geom_pointrange(data=PDXfile, aes(x=library_label, y=coefficient, ymin=lowerconfint, ymax=upperconfint, color=group),size=0.2) +
  geom_errorbar(data=PDXfile, aes(x=library_label, y=coefficient, ymin=lowerconfint, ymax=upperconfint, color=group),width=0.4) +
  geom_point(data=PDXfile, aes(x=library_label, y=coefficient, color=fillcol), size=0.2) +
  ylab("Pearson correlation coefficient") +
  geom_hline(yintercept = 0, lty='dashed') +
  geom_pointrange() +
  scale_color_manual(values = colorlist) +
  scale_x_discrete(limits = levels(PDXfile$library_label), labels = PDXfile$shortlibid) +
  scale_y_continuous(limits = c(-0.6,0.9)) +
  theme_bw() +
  coord_flip() +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_text(size=6),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.line = element_line(size = 0.5, colour = "black"),
        panel.grid.minor = element_blank(),
        legend.position = "none",
        panel.border = element_blank())

primaryfile$fillcol = 'white'
primaryfile$fillcol[which(primaryfile$signif == 1)] = primaryfile$group[which(primaryfile$signif == 1)]
primaryfile$library_label = factor(primaryfile$library_label,levels = unique(primaryfile$library_label))

figprimary = ggplot() +
  geom_pointrange(data=primaryfile, aes(x=library_label, y=coefficient, ymin=lowerconfint, ymax=upperconfint, color=group),size=0.2) +
  geom_errorbar(data=primaryfile, aes(x=library_label, y=coefficient, ymin=lowerconfint, ymax=upperconfint, color=group),width=0.4) +
  geom_point(data=primaryfile, aes(x=library_label, y=coefficient, color=fillcol), size=0.2) +
  ylab("Pearson correlation coefficient") +
  geom_hline(yintercept = 0, lty='dashed') +
  geom_pointrange() +
  scale_color_manual(values = colorlist) +
  scale_x_discrete(limits = levels(primaryfile$library_label), labels = primaryfile$shortlibid) +
  scale_y_continuous(limits = c(-0.6,0.9)) +
  theme_bw() +
  coord_flip() +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_text(size=6),
        axis.text.x = element_text(size=6),
        axis.title.x = element_text(size=7.5),
        axis.line = element_line(size = 0.5, colour = "black"),
        panel.grid.minor = element_blank(),
        legend.position = "none",
        panel.border = element_blank())


print(figcell_line)

print(plot_grid(figPDX,figprimary, nrow = 2, rel_heights = c(0.65,0.35), align = 'v'))





```

***

# Figure 3

```{r eval=T, echo=F}

# Import the master result file across all cells
resultfile = read.csv(paste0('../data/processed_data/final_master.tsv'), header=T, sep='\t', stringsAsFactors=F)

# Load clone label file
clonelabels <- read.table(paste0('../data/processed_data/clonelabels.tsv'), header=T, sep='\t', stringsAsFactors=F)

## Load SA906a file: can also import SA906a_combined.RData file from Zenodo (https://zenodo.org/records/10498240)
# load("SA906a_combined.RData")
SA906a_CNbins <- read.table(paste0('../data/processed_data/SA906a_CNbins.tsv'), header=T, sep='\t', stringsAsFactors=F)
SA906a_clonelabels <- clonelabels[which(clonelabels$sample_id == 'SA906a'),]
SA906a_allclones <- c("2N D","4N D","2N C","4N C","2N G","4N G","2N A","4N A")

## Load OV-081 file: can also import OV081_combined.RData file from Zenodo (https://zenodo.org/records/10498240)
# load("OV081_combined.RData")
OV081_CNbins <- read.table(paste0('../data/processed_data/OV081_CNbins.tsv'), header=T, sep='\t', stringsAsFactors=F)
OV081_clonelabels <- clonelabels[which(clonelabels$sample_id == 'OV-081'),]
OV081_allclones <- c("2N C","4N C","2N B","2N A")

```


## 3a

```{r panel_3a,fig.height=2,fig.width=3,eval=T}

filtresultfile = resultfile[which(resultfile$ploidylab %in% c(2,4)),]
filtresultfile$ploidylab = factor(filtresultfile$ploidylab, levels=c(2,4))
keepthese = filtresultfile %>% group_by(library_id) %>% distinct(library_id,ploidylab) %>% mutate(n=n()) %>% filter(n == 2) %>% distinct(library_id)
filtresultfile = filtresultfile[which(filtresultfile$library_id %in% keepthese$library_id),]

ploidyinfo = filtresultfile %>% group_by(library_id,ploidylab) %>% mutate(n = n()) %>% distinct(library_id,n)

res_summary = filtresultfile %>% group_by(library_id) %>% mutate(medtet=median(copynumber[ploidylab==4]), meddip=median(copynumber[ploidylab==2]), sdtet=sd(copynumber[ploidylab==4]), sddip=sd(copynumber[ploidylab==2]), diffmed=median(copynumber[ploidylab==4])-median(copynumber[ploidylab==2]), diffsd=sd(copynumber[ploidylab==4])-sd(copynumber[ploidylab==2])) %>% distinct(library_id,meddip,medtet,sddip,sdtet,diffmed,diffsd)

# Separate the diploid and tetraploid cells
diploidinfo = ploidyinfo[which(ploidyinfo$ploidylab == 2),]
tetraploidinfo = ploidyinfo[which(ploidyinfo$ploidylab == 4),]

# Obtain the sample size
res_summary$ndip = diploidinfo$n[match(res_summary$library_id,diploidinfo$library_id)]
res_summary$ntet = tetraploidinfo$n[match(res_summary$library_id,tetraploidinfo$library_id)]

# Filter for libraries with at least 15 cells in both groups
res_summary = res_summary[which((res_summary$ndip >= 15) & (res_summary$ntet >= 15)),]
table(res_summary$medtet > res_summary$meddip*2)

# Comparing the MNR
filtresultfile = filtresultfile[which(filtresultfile$library_id %in% res_summary$library_id),]

signiflibs = c()
for (eachlib in unique(filtresultfile$library_id)){
  currlib = filtresultfile[which(filtresultfile$library_id == eachlib),]
  tetdata = data.frame("4N",currlib$MNR[which(currlib$ploidylab == 4)])
  colnames(tetdata) = c("ploidy","MNR")
  dipdata = data.frame("2N",currlib$MNR[which(currlib$ploidylab == 2)])
  colnames(dipdata) = c("ploidy","MNR")
  currdata = rbind(tetdata,dipdata)
  
  testres <- wilcox.test(currdata$MNR ~ currdata$ploidy)
  if (testres$p.value < 0.05){
    signiflibs = append(signiflibs,eachlib)
  }
}

filtresultfile = filtresultfile[order(as.numeric(filtresultfile$copynumber), decreasing = T),]
filtresultfile$cell_id = factor(filtresultfile$cell_id, levels = unique(filtresultfile$cell_id))
filtresultfile <- filtresultfile[which(filtresultfile$ploidylab %in% c(2,4)),]
filtresultfile$newploidylab = as.character(filtresultfile$ploidylab)
filtresultfile$newploidylab[which(filtresultfile$newploidylab == 4)] = "4N"
filtresultfile$newploidylab[which(filtresultfile$newploidylab == 2)] = "2N"
filtresultfile$newploidylab = factor(filtresultfile$newploidylab, levels = c("2N","4N"))

my_comparisons <- list(c('2N','4N'))

fig3a <- ggplot(filtresultfile, aes(x=newploidylab, y=log2(copynumber))) + geom_violin(aes(fill=newploidylab)) + geom_boxplot(outlier.size = -1, width=0.2) +
  scale_fill_manual(name="Ploidy",values=c("#5074AF","#F1AF31")) + theme_classic() + xlab("Ploidy") + scale_x_discrete(expand = c(0.04,0.04)) +
  scale_y_continuous(expand = c(0,0), limits=c(8,14)) +
  stat_compare_means(comparisons = my_comparisons, label.y = 12, method = "wilcox.test", size = 4, label = 'p.signif') +
  ylab(expression('log'[2]~'(mtDNA copy number)')) +
  theme(axis.title.y = element_text(size=8),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.key.size = unit(0.4, 'cm'))

fig3a



```


## 3b

```{r panel_3b,fig.height=2,fig.width=4,eval=T}

## tetraploid comparison
cell_linefile = resultfile[grep('184hTERT',resultfile$group),]
cell_linefile = cell_linefile[which(cell_linefile$ploidylab %in% c(2,4)),]
cell_linefile$ploidylab = factor(cell_linefile$ploidylab, levels=c(2,4))

# Diploid and tetraploid analysis with marginal histogram
# Count the number of diploid and tetraploid cells
filtfinalres_n1 = cell_linefile %>% group_by(group, ploidylab) %>% filter(ploidylab == 4) %>% dplyr::mutate(n1 = n()) %>% filter(n1 >= 10) %>% dplyr::select(group,ploidylab,n1) %>% distinct()
filtfinalres_n2 = cell_linefile %>% group_by(group, ploidylab) %>% filter(ploidylab == 2) %>% dplyr::mutate(n2 = n()) %>% filter(n2 >= 10) %>% dplyr::select(group,ploidylab,n2) %>% distinct()

# Keep those groups that have both diploid and tetraploid cells
filtfinalres_n1 = filtfinalres_n1[which(filtfinalres_n1$group %in% intersect(filtfinalres_n2$group, filtfinalres_n1$group)),]
filtfinalres_n2 = filtfinalres_n2[which(filtfinalres_n2$group %in% intersect(filtfinalres_n1$group, filtfinalres_n2$group)),]

# Concatenating the sample size for diploid and tetraploid cells
filtfinalres_n = data.frame(intersect(filtfinalres_n2$group, filtfinalres_n1$group))
filtfinalres_n$n = paste0("n=",filtfinalres_n1$n1,",n=",filtfinalres_n2$n2)
colnames(filtfinalres_n) = c("group","n")

# Reassign the group label to include sample size
cell_linefile = cell_linefile[which(cell_linefile$group %in% intersect(cell_linefile$group, filtfinalres_n$group)),]
cell_linefile$group = paste0(cell_linefile$group, "\n", filtfinalres_n$n[match(cell_linefile$group, filtfinalres_n$group)])

cell_lines <- cell_linefile[,c('cell_id','ploidylab','MNR','group')]
colnames(cell_lines) = c('cell_id','ploidylab','MNR','group')

cell_lines$group = factor(cell_lines$group, levels = c("184hTERT\nn=350,n=3475",
                                                       "184hTERT(TP53-/-)(SA906a)\nn=1234,n=3376",
                                                       "184hTERT(TP53-/-)(SA906b)\nn=833,n=5555",
                                                       "184hTERT(TP53-/-)(SA1101a)\nn=1012,n=5290",
                                                       "184hTERT(TP53-/-)(SA1101b)\nn=1071,n=10503",
                                                       "184hTERT(BRCA2+/-;TP53-/-)(SA1188)\nn=101,n=1876",
                                                       "184hTERT(BRCA2-/-;TP53-/-)(SA1055)\nn=199,n=264",
                                                       "184hTERT(BRCA2-/-;TP53-/-)(SA1056)\nn=25,n=549",
                                                       "184hTERT(BRCA1+/-;TP53-/-)(SA1292)\nn=11,n=446",
                                                       "184hTERT(BRCA1-/-;TP53-/-)(SA1054)\nn=216,n=131"))

## cell_lines
fig3b <- ggplot(cell_lines, aes(x=group, y=MNR)) + geom_violin(aes(fill=ploidylab), position = position_dodge(width = 0.6), trim = F, linewidth=0.2) +
  scale_y_continuous(limits=c(0,1300),expand=c(0,0)) + #stat_compare_means(aes(label = paste0("p = ", after_stat(p.format))), label.y = 1200,size=2) +
  geom_boxplot(aes(group=interaction(group,ploidylab)), fill="white", position = position_dodge(width = 0.6), width=0.1, linewidth=0.2, outlier.size = -1) + theme_classic() + scale_fill_manual(name="Ploidy",values=c("#5074AF","#F1AF31")) +
  ylab("MNR")

fig3b <- fig3b + theme(axis.text.y = element_text(size=7),
                       axis.text.x = element_blank(), # element_text(size=7,angle = 45, vjust=0.5, hjust=0.5),
                       axis.title = element_text(size=7),
                       axis.title.x = element_blank(),
                       legend.position='none',
                       axis.line = element_line(linewidth=0.4),
                       axis.ticks = element_line(linewidth=0.4),
                       strip.background = element_blank())

fig3b_legend <- as_ggplot(get_legend(fig3b))

fig3b <- fig3b + theme(legend.position='none')

comparisonresult <- cell_lines %>% 
  group_by(group,ploidylab) %>%
  dplyr::summarise(
    n=n(), # number of cells
    median=median(MNR), # median MNR
    sd=sd(MNR) # standard deviation of MNR
  ) %>%
  mutate(se=sd/sqrt(n)) %>% # standard error of MNR calculated by dividing standard deviation of MNR by the sqrt of sample size
  group_by(group) %>%
  filter(all(c(2, 4) %in% ploidylab))

diffresult <- comparisonresult %>%
  dplyr::mutate(diffVal = (median[ploidylab == 4] - median[ploidylab == 2])/median[ploidylab == 2] * 100) %>%
  dplyr::mutate(sediffVal = abs(median[ploidylab == 4] / median[ploidylab == 2]) * sqrt((se[ploidylab == 4]^2) / (median[ploidylab == 4]^2) + (se[ploidylab == 2]^2) / (median[ploidylab == 2]^2)) * 100) %>%
  group_by(group) %>%
  dplyr::mutate(uci = diffVal + 1.65 * sediffVal, lci = diffVal - 1.65 * sediffVal) %>%
  dplyr::select(c("group","diffVal","sediffVal","uci","lci")) %>%
  distinct()

diffresult$Type = 'cell_lines'
diffresult$label = ''
diffresult$label[which(diffresult$diffVal == max(diffresult$diffVal))] = strsplit(as.character(diffresult$group[which(diffresult$diffVal == max(diffresult$diffVal))]),'\n')[[1]][1]
side_fig3b <- ggplot(diffresult, aes(x=Type,y=diffVal)) + geom_boxplot(outlier.size = -1) + geom_jitter(size=0.7,width=0.2) +
  scale_y_continuous(limits = c(-5,40), expand=c(0.01,0.01)) +
  ylab("% Change") + theme_bw() + geom_hline(yintercept = 0) +
  theme(axis.text.x = element_blank(),
        axis.title.y = element_text(size=7),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank())

## Cell lines
fig3b_cell_lines <- plot_grid(fig3b, side_fig3b, nrow=1, align = 'h', rel_widths = c(0.82,0.18))

fig3b_cell_lines


```


## 3c

```{r panel_3c,fig.height=2,fig.width=4,eval=T}

## tetraploid comparison
filtresultfile = resultfile[-which((resultfile$clone_UMAP == 'nonmal') | (resultfile$group == 'GM18507')),]
filtresultfile = filtresultfile[which(filtresultfile$ploidylab %in% c(2,4)),]
filtresultfile$ploidylab = factor(filtresultfile$ploidylab, levels=c(2,4))

PDXfile = filtresultfile[grep('PDX',filtresultfile$group),]
untreated = PDXfile[grep('untreated',PDXfile$group),]
PDXfile = PDXfile[-grep('treated',PDXfile$group),]
PDXfile = rbind(PDXfile, untreated)
PDXfile$group = gsub('untreated ','',PDXfile$group)
PDXfile$group = gsub('PDX','',PDXfile$group)
PDXfile$group = gsub('TNBC ','',PDXfile$group)
PDXfile$group = gsub("[()]", "", PDXfile$group)

# Diploid and tetraploid analysis with marginal histogram
# Count the number of diploid and tetraploid cells
filtfinalres_n1 = PDXfile %>% group_by(group, ploidylab) %>% filter(ploidylab == 4) %>% dplyr::mutate(n1 = n()) %>% filter(n1 >= 10) %>% dplyr::select(group,ploidylab,n1) %>% distinct()
filtfinalres_n2 = PDXfile %>% group_by(group, ploidylab) %>% filter(ploidylab == 2) %>% dplyr::mutate(n2 = n()) %>% filter(n2 >= 10) %>% dplyr::select(group,ploidylab,n2) %>% distinct()

# Keep those groups that have both diploid and tetraploid cells
filtfinalres_n1 = filtfinalres_n1[which(filtfinalres_n1$group %in% intersect(filtfinalres_n2$group, filtfinalres_n1$group)),]
filtfinalres_n2 = filtfinalres_n2[which(filtfinalres_n2$group %in% intersect(filtfinalres_n1$group, filtfinalres_n2$group)),]

# Concatenating the sample size for diploid and tetraploid cells
filtfinalres_n = data.frame(intersect(filtfinalres_n2$group, filtfinalres_n1$group))
filtfinalres_n$n = paste0("n=",filtfinalres_n1$n1,",n=",filtfinalres_n2$n2)
colnames(filtfinalres_n) = c("group","n")

# Reassign the group label to include sample size
PDXfile = PDXfile[which(PDXfile$group %in% intersect(PDXfile$group, filtfinalres_n$group)),]
PDXfile$group = paste0(PDXfile$group, "\n", filtfinalres_n$n[match(PDXfile$group, filtfinalres_n$group)])

PDXes <- PDXfile[,c('cell_id','ploidylab','MNR','group')]
colnames(PDXes) = c('cell_id','ploidylab','MNR','group')

primary <- filtresultfile[grep('primary',filtresultfile$group),]

# Diploid and tetraploid analysis with marginal histogram
# Count the number of diploid and tetraploid cells
filtfinalres_n1 = primary %>% group_by(group, ploidylab) %>% filter(ploidylab == 4) %>% dplyr::mutate(n1 = n()) %>% filter(n1 >= 10) %>% dplyr::select(group,ploidylab,n1) %>% distinct()
filtfinalres_n2 = primary %>% group_by(group, ploidylab) %>% filter(ploidylab == 2) %>% dplyr::mutate(n2 = n()) %>% filter(n2 >= 10) %>% dplyr::select(group,ploidylab,n2) %>% distinct()

# Keep those groups that have both diploid and tetraploid cells
filtfinalres_n1 = filtfinalres_n1[which(filtfinalres_n1$group %in% intersect(filtfinalres_n2$group, filtfinalres_n1$group)),]
filtfinalres_n2 = filtfinalres_n2[which(filtfinalres_n2$group %in% intersect(filtfinalres_n1$group, filtfinalres_n2$group)),]

# Concatenating the sample size for diploid and tetraploid cells
filtfinalres_n = data.frame(intersect(filtfinalres_n2$group, filtfinalres_n1$group))
filtfinalres_n$n = paste0("n=",filtfinalres_n1$n1,",n=",filtfinalres_n2$n2)
colnames(filtfinalres_n) = c("group","n")

# Reassign the group label to include sample size
primary = primary[which(primary$group %in% intersect(primary$group, filtfinalres_n$group)),]
primary$group = paste0(primary$group, "\n", filtfinalres_n$n[match(primary$group, filtfinalres_n$group)])

primary <- primary[,c('cell_id','ploidylab','MNR','group')]
finalfile = rbind(PDXes,primary)

finalfile$group = factor(finalfile$group,levels = c("SA1035\nn=178,n=3040",
                                                    "TNBC primary(SA1135)\nn=76,n=73",
                                                    "SA1142\nn=33,n=2475",
                                                    "SA501\nn=121,n=2264",
                                                    "SA535\nn=41,n=1861",
                                                    "SA609\nn=55,n=4689",
                                                    "HGSC primary(OV-022)\nn=109,n=1382",
                                                    "HGSC primary(OV-081)\nn=456,n=274"))

## PDX and primary
fig3c <- ggplot(finalfile, aes(x=group, y=MNR)) + geom_violin(aes(fill=ploidylab), position = position_dodge(width = 0.6), trim = F, linewidth=0.2) +
  scale_y_continuous(limits=c(0,1300),expand=c(0,0)) + #stat_compare_means(aes(label = paste0("p = ", after_stat(p.format))), label.y = 1200,size=2) +
  geom_boxplot(aes(group=interaction(group,ploidylab)), fill="white", position = position_dodge(width = 0.6), width=0.1, linewidth=0.2, outlier.size = -1) +
  theme_classic() + scale_fill_manual(name="Ploidy",values=c("#5074AF","#F1AF31")) + ylab("MNR")

fig3c <- fig3c + theme(axis.text.y = element_text(size=7),
                       axis.text.x = element_blank(), # element_text(size=7,angle = 45, vjust=0.5, hjust=0.5),
                       axis.title = element_text(size=7),
                       axis.title.x = element_blank(),
                       legend.title=element_text(size=3.5),
                       legend.text=element_text(size=3.5),
                       legend.key.size = unit(0.3, 'cm'), #change legend key size
                       legend.key.height = unit(0.3, 'cm'), #change legend key height
                       legend.key.width = unit(0.3, 'cm'), #change legend key width
                       axis.line = element_line(size=0.4),
                       axis.ticks = element_line(size=0.4),
                       strip.background = element_blank())

fig3c_legend <- as_ggplot(get_legend(fig3c))

fig3c <- fig3c + theme(legend.position='none')

comparisonresult <- finalfile %>% 
  group_by(group,ploidylab) %>%
  dplyr::summarise(
    n=n(), # number of cells
    median=median(MNR), # median MNR
    sd=sd(MNR) # standard deviation of MNR
  ) %>%
  mutate(se=sd/sqrt(n)) %>% # standard error of MNR calculated by dividing standard deviation of MNR by the sqrt of sample size
  group_by(group) %>%
  filter(all(c(2, 4) %in% ploidylab))

diffresult <- comparisonresult %>%
  dplyr::mutate(diffVal = (median[ploidylab == 4] - median[ploidylab == 2])/median[ploidylab == 2] * 100) %>%
  dplyr::mutate(sediffVal = abs(median[ploidylab == 4] / median[ploidylab == 2]) * sqrt((se[ploidylab == 4]^2) / (median[ploidylab == 4]^2) + (se[ploidylab == 2]^2) / (median[ploidylab == 2]^2)) * 100) %>%
  group_by(group) %>%
  dplyr::mutate(uci = diffVal + 1.65 * sediffVal, lci = diffVal - 1.65 * sediffVal) %>%
  dplyr::select(c("group","diffVal","sediffVal","uci","lci")) %>%
  distinct()

diffresult$Type = 'tumor'
diffresult$label = ''
diffresult$label[which(diffresult$diffVal == max(diffresult$diffVal))] = strsplit(as.character(diffresult$group[which(diffresult$diffVal == max(diffresult$diffVal))]),'\n')[[1]][1]
side_fig3c <- ggplot(diffresult, aes(x=Type,y=diffVal)) + geom_boxplot(outlier.size = -1) + geom_jitter(size=0.7,width=0.2) +
  scale_y_continuous(limits = c(-5,18), expand=c(0.01,0.01)) +
  ylab("% Change") + theme_bw() + geom_hline(yintercept = 0) +
  theme(axis.title.y = element_text(size=7),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank())

## PDX and primary
fig3c_tumor <- plot_grid(fig3c, side_fig3c, nrow=1, align = 'h', rel_widths = c(0.78,0.22))
fig3c_tumor


```


## 3d

```{r panel_3d,fig.height=5,fig.width=4,eval=T}

## Iterate through the clones and plot the nuDNA copy number profiles
clonefigslist = c()
for (eachclone in SA906a_allclones){
  figlab <- ggdraw() + draw_label(paste0('n=',SA906a_clonelabels$ncells[which(SA906a_clonelabels$clone_id == eachclone)]), size=12, x = 0.95, angle = 90, vjust = -0.1, hjust = 0) + theme(plot.margin = margin(0, 0, 0, 0))
  fig <- plotCNprofile(CNbins = SA906a_CNbins[which(SA906a_CNbins$cell_id == eachclone),], maxCN = 8, pointsize = 0.4, legend.position = "none") +
    theme(axis.text.y = element_text(size=8),
          axis.text.x = element_text(size=10),
          axis.title = element_blank())
  assign(paste0('fig',eachclone), plot_grid(fig, figlab, rel_widths = c(1,0.03)))
  clonefigslist = append(clonefigslist,paste0('fig',eachclone))
}

fig3d <- plot_grid(plotlist=mget(rev(clonefigslist)), nrow = length(SA906a_allclones))
fig3d

```

## 3e

```{r panel_3e,fig.height=2.5,fig.width=2,eval=T}

## MNR plot and nuDNA plot across tetraploid and diploid for SA906a
SA906a_mtDNAcopynumber = resultfile[which(resultfile$group == "184hTERT(TP53-/-)(SA906a)"),]
SA906a_mtDNAcopynumber$clone_id = paste0(SA906a_mtDNAcopynumber$ploidylab,'N ',SA906a_mtDNAcopynumber$clone_UMAP)
SA906a_mtDNAcopynumber = SA906a_mtDNAcopynumber[which(SA906a_mtDNAcopynumber$clone_id %in% SA906a_allclones),]
SA906a_mtDNAcopynumber$clone_id = factor(SA906a_mtDNAcopynumber$clone_id,levels = SA906a_allclones)

log2FC = c()
for (eachclone in c('A','C','D','G')){
  log2FC = append(log2FC,
                  round(log2(median(SA906a_mtDNAcopynumber$MNR[which(SA906a_mtDNAcopynumber$clone_id == paste0('4N ',eachclone))])/
                               median(SA906a_mtDNAcopynumber$MNR[which(SA906a_mtDNAcopynumber$clone_id == paste0('2N ',eachclone))])), digits = 4))
}
log2FC_2N_AD = round(log2(median(SA906a_mtDNAcopynumber$MNR[which(SA906a_mtDNAcopynumber$clone_id == '2N A')])/
                      median(SA906a_mtDNAcopynumber$MNR[which(SA906a_mtDNAcopynumber$clone_id == '2N D')])),digits = 2)

my_comparisons <- list(c("2N A","2N D"),c("2N A","4N A"),c("2N C","4N C"),c("2N D","4N D"),c("2N G","4N G"))

# Plot the mtDNA copy number
fig3e = ggplot(SA906a_mtDNAcopynumber, aes(x=clone_id, y=MNR, fill = clone_id)) + geom_violin(width = 0.8, trim=FALSE) +
  geom_boxplot(fill="white", width=0.08, outlier.size=-1) + theme_classic() +
  stat_compare_means(aes(label = after_stat(p.signif)),comparisons = my_comparisons, label.y = c(1100, 970, 970, 970, 970), size = 2, method = 'wilcox.test') +
  annotate("text", label = paste0("log2(FC)=",log2FC), x = rev(seq(from=1.5,by=2,to=1.5+2*(length(SA906a_allclones)/2-1))), y = rep(1500,length(SA906a_allclones)/2), size = 2) +
  annotate("text", label = paste0("log2(FC)=",log2FC_2N_AD), x = 4.9, y = 1600, size = 2) +
  theme(axis.text = element_text(size=8),
        axis.text.y = element_blank(),
        axis.title = element_blank(),
        panel.grid.major.y = element_line(colour = "grey",linetype="dashed",size=0.5),
        legend.position = "none") +
  scale_y_continuous(expand = c(0,0), limits = c(0,1800)) + xlab(label = "Clones") + ylab(label = "MNR") +
  coord_flip() +
  scale_fill_manual(values = c("4N A" = "#7A4644", "2N A" = "#7A4644",
                               "4N C" = "#A56937", "2N C" = "#A56937",
                               "4N D" = "#843D8D", "2N D" = "#843D8D",
                               "4N G" = "#859071", "2N G" = "#859071"))

fig3e


```


## 3f

```{r panel_3f,fig.height=5,fig.width=4,eval=T}

## nuDNA plot across tetraploid and diploid for OV-081
## Iterate through the clones and plot the nuDNA copy number profiles
clonefigslist = c()
for (eachclone in OV081_allclones){
  figlab <- ggdraw() + draw_label(paste0('n=',OV081_clonelabels$ncells[which(OV081_clonelabels$clone_id == eachclone)]), size=12, x = 0.95, angle = 90, vjust = -0.1, hjust = 0) + theme(plot.margin = margin(0, 0, 0, 0))
  assign(paste0('figlab',eachclone), figlab)
  fig <- plotCNprofile(CNbins = OV081_CNbins[which(OV081_CNbins$cell_id == eachclone),], maxCN = 8, pointsize = 0.4, legend.position = "none") +
    theme(axis.text.x = element_text(size=10),
          axis.text.y = element_text(size=7),
          axis.title = element_blank())
  assign(paste0('fig',eachclone), fig)
  clonefigslist = append(clonefigslist,paste0('fig',eachclone))
}

eachclone = '4N C'
figlab <- ggdraw() + draw_label(paste0('n=',OV081_clonelabels$ncells[which(OV081_clonelabels$clone_id == eachclone)]), size=12, x = 0.95, angle = 90, vjust = -0.1, hjust = 0) + theme(plot.margin = margin(0, 0, 0, 0))
assign(paste0('figlab',eachclone), figlab)
fig <- plotCNprofile(CNbins = OV081_CNbins[which(OV081_CNbins$cell_id == eachclone),], maxCN = 10, pointsize = 0.4, legend.position = "none") +
  theme(axis.text.x = element_text(size=10),
        axis.text.y = element_text(size=7),
        axis.title = element_blank())
assign(paste0('fig',eachclone), fig)

nuDNAfig <- plot_grid(plotlist=mget(rev(clonefigslist)), nrow = (length(OV081_allclones)), align = 'v')

nuDNAfiglab <- plot_grid(plotlist=mget(rev(gsub('fig','figlab',clonefigslist))), nrow = (length(OV081_allclones)), align = 'v')

fig3f <- plot_grid(nuDNAfig, nuDNAfiglab, nrow = 1, rel_widths = c(1,0.03))
fig3f

```

## 3g

```{r panel_3g,fig.height=2,fig.width=3,eval=T}

# Plot the mtDNA copy number
## MNR plot and nuDNA plot across tetraploid and diploid for OV-081
OV081_mtDNAcopynumber = resultfile[which(resultfile$group == "HGSC primary(OV-081)"),]
OV081_mtDNAcopynumber$clone_id = paste0(OV081_mtDNAcopynumber$ploidylab,'N ',OV081_mtDNAcopynumber$clone_UMAP)
OV081_mtDNAcopynumber = OV081_mtDNAcopynumber[which(OV081_mtDNAcopynumber$clone_id %in% OV081_allclones),]
OV081_mtDNAcopynumber$clone_id = factor(OV081_mtDNAcopynumber$clone_id,levels = OV081_allclones)

log2FC = round(log2(median(OV081_mtDNAcopynumber$MNR[which(OV081_mtDNAcopynumber$clone_id == '4N C')])/
                               median(OV081_mtDNAcopynumber$MNR[which(OV081_mtDNAcopynumber$clone_id == '2N C')])), digits = 4)

log2FC_AC = round(log2(median(OV081_mtDNAcopynumber$MNR[which(OV081_mtDNAcopynumber$clone_id == '2N C')])/
                               median(OV081_mtDNAcopynumber$MNR[which(OV081_mtDNAcopynumber$clone_id == '2N A')])), digits = 4)

my_comparisons <- list(c("2N A","2N C"),c("2N C","4N C"))

# Plot the mtDNA copy number
mtDNAfig = ggplot(OV081_mtDNAcopynumber, aes(x=clone_id, y=MNR, fill = clone_id)) + geom_violin(width = 0.8, trim=FALSE) +
  geom_boxplot(fill="white", width=0.08, outlier.size=-1) + theme_classic() +
  stat_compare_means(aes(label = after_stat(p.signif)),comparisons = my_comparisons, label.y = c(1100, 970), size = 2, method = 'wilcox.test') +
  annotate("text", label = paste0("log2(FC)=",log2FC), x = 1.5, y = 1500, size = 2) +
  annotate("text", label = paste0("log2(FC)=",log2FC_AC), x = 2.5, y = 1500, size = 2) +
  theme(axis.text = element_text(size=8),
        axis.text.y = element_blank(),
        axis.title = element_blank(),
        panel.grid.major.y = element_line(colour = "grey",linetype="dashed",size=0.5),
        legend.position = "none") +
  scale_y_continuous(expand = c(0,0), limits = c(0,1800)) + xlab(label = "Clones") + ylab(label = "MNR") +
  coord_flip() +
  scale_fill_manual(values = c("2N A" = "#7A4644", "2N B" = "#89C4B5",
                               "4N C" = "#C9B2D4", "2N C" = "#C9B2D4"))

mtDNAfig


```


***

# Figure 4


```{r eval=T, echo=F}

OV022_scRNA <- read.table(paste0('../data/processed_data/OV022_scRNA.tsv'), header=T, sep='\t', stringsAsFactors=F)
OV081_scRNA <- read.table(paste0('../data/processed_data/OV081_scRNA.tsv'), header=T, sep='\t', stringsAsFactors=F)

tumor_fgseafile = read.csv(paste0('../data/processed_data/tumor_fgseacombined.tsv'), header=T, sep='\t', stringsAsFactors=F)

```


## 4a

```{r panel_4a,fig.height=3,fig.width=4,eval=T}

## Plot UMAP
fig4a <- ggplot(OV022_scRNA, aes(x=UMAP_1,y=UMAP_2,color=clonealign_id)) + geom_point(size=0.5) + 
  scale_color_manual(values=customcolors$clones) +scale_x_continuous(limits= c(-6,-2)) + scale_y_continuous(limits= c(1,4)) + theme_minimal() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(size = 14),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=8),
        legend.title = element_blank())
fig4a

```

## 4b

```{r panel_4b,fig.height=3,fig.width=4,eval=T}

## Import final master file
filtresultfile = resultfile[which(resultfile$group == "HGSC primary(OV-022)"),]
filtresultfile = filtresultfile[-which(is.na(filtresultfile$treealign_id)),]
filtresultfile = filtresultfile[which(filtresultfile$treealign_id %in% sort(unique(filtresultfile$treealign_id))),]

fig4b <- ggplot(filtresultfile, aes(x=reorder(treealign_id,-MNR),y=MNR)) + geom_violin(aes(fill=treealign_id),trim = F) + geom_boxplot(width=0.2,outlier.size = -1) + theme_classic() +
  scale_y_continuous(limits = c(0,1600),expand = c(0,0)) + scale_fill_manual(values=customcolors$clones) +
  theme(axis.text = element_text(size=8),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=8),
        legend.title = element_blank(),
        legend.position = 'none')
fig4b

```


## 4c

```{r panel_4c,fig.height=3,fig.width=4,eval=T}

## Plot CO2 gene expression
fig4c <- ggplot(OV022_scRNA, aes(x=UMAP_1,y=UMAP_2,color=MT.CO2)) + geom_point(size=0.5) +
  scale_color_gradientn(colours = c("blue", "blue", "white", "red"), values = rescale(x = c(0, 2.5, 3.8, 5)), name = 'Expression') +
  ggtitle('MT-CO2') + scale_x_continuous(limits= c(-6,-2)) + scale_y_continuous(limits= c(1,4)) + theme_minimal() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(size = 14),
        legend.key.size = unit(0.6, 'cm'),
        legend.text = element_text(size=8))
fig4c

```


## 4d

```{r panel_4d,fig.height=3,fig.width=4,eval=T}

## Plot UMAP
fig4d <- ggplot(OV081_scRNA, aes(x=UMAP_1,y=UMAP_2,color=clonealign_id)) + geom_point(size=0.5) + 
  scale_color_manual(values=customcolors$clones) +scale_x_continuous(limits=c(-10.5,-6.5)) + scale_y_continuous(limits = c(10.5,14)) + theme_minimal() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(size = 14),
        legend.key.size = unit(1, 'cm'),
        legend.text = element_text(size=8),
        legend.title = element_blank())
fig4d

```


## 4e

```{r panel_4e,fig.height=3,fig.width=4,eval=T}

## Import final master file
filtresultfile = resultfile[which(resultfile$group == "HGSC primary(OV-022)"),]
filtresultfile = filtresultfile[-which(is.na(filtresultfile$treealign_id)),]
filtresultfile = filtresultfile[which(filtresultfile$treealign_id %in% sort(unique(filtresultfile$treealign_id))),]

fig4e <- ggplot(filtresultfile, aes(x=reorder(treealign_id,-MNR),y=MNR)) + geom_violin(aes(fill=treealign_id),trim = F) + geom_boxplot(width=0.2,outlier.size = -1) + theme_classic() +
  scale_y_continuous(limits = c(0,1600),expand = c(0,0)) + scale_fill_manual(values=customcolors$clones) +
  theme(axis.text = element_text(size=8),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=8),
        legend.title = element_blank(),
        legend.position = 'none')

fig4e

```


## 4f

```{r panel_4f,fig.height=3,fig.width=4,eval=T}

## Plot ND3 gene expression
fig4f <- ggplot(OV081_scRNA, aes(x=UMAP_1,y=UMAP_2,color=MT.ND3)) + geom_point(size=0.5) +
  scale_color_gradientn(colours = c("blue", "blue", "white", "red"), values = rescale(x = c(0, 2.5, 3.8, 5)), name = 'Expression') +
  ggtitle('MT-ND3') + scale_x_continuous(limits=c(-10.5,-6.5)) + scale_y_continuous(limits = c(10.5,14)) + theme_minimal() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(size = 14),
        legend.key.size = unit(0.6, 'cm'),
        legend.text = element_text(size=8))

fig4f


```


## 4g

```{r panel_4g,fig.height=2.5,fig.width=2.5,eval=T}

summarize_signature <- function(m) {
  mid <- median(m$stat)
  n <- sum(abs(m$stat) > 2) #/ nrow(m)
  m <- m[order(stat,decreasing=F),]
  m$n <- n
  m$mid <- mid
  m$pos <- ((1:nrow(m))/nrow(m))
  m
}
tumor_fgseafile.dt <- adt(tumor_fgseafile)
tumor_fgseafile.dt <- tumor_fgseafile.dt[,summarize_signature(.SD), by=pathway]
tumor_fgseafile.dt <- tumor_fgseafile.dt[order(mid,pos,decreasing=F),]
tumor_fgseafile.dt <- tumor_fgseafile.dt[n >= 2,]
tumor_fgseafile.dt$pathway <- factor(tumor_fgseafile.dt$pathway, levels=unique(tumor_fgseafile.dt$pathway))
tumor_fgseafile.dt$label <- tumor_fgseafile.dt$pathway
tumor_fgseafile.dt$pos <- as.integer(tumor_fgseafile.dt$pathway) + tumor_fgseafile.dt$pos - 0.5
info <- tumor_fgseafile.dt[!duplicated(pathway),]

tumor_fgseafile.dt$n = factor(tumor_fgseafile.dt$n, levels = sort(unique(tumor_fgseafile.dt$n)))

fig4g <- ggplot(tumor_fgseafile.dt, aes(x=pos,y=stat)) +
  scale_y_continuous(limits=c(-10,10)) + 
  geom_bar(data=info, stat='identity',aes(x=pathway, y=mid),color='black',fill='#4675b4') +
  theme_classic() +
  coord_flip() +
  theme(axis.text.x=element_text(size=5),
        axis.text.y = element_text(size=6),
        axis.title = element_text(size=8),
        legend.key.size = unit(0.4,'cm'),
        legend.key.height = unit(0.4,'cm'),
        legend.key.width = unit(0.4,'cm'),
        legend.title = element_text(size=5),
        legend.text = element_text(size=5),
        legend.position = 'bottom') + 
  labs(x=NULL,y=expression('-log10(Q-value)*\u394(NES)'))
fig4g


```


## 4h

```{r panel_4h,fig.height=3,fig.width=4,eval=T}

## Plot PROGENy Hypoxia signature
fig4h <- ggplot(OV022_scRNA, aes(x=UMAP_1,y=UMAP_2,color=Hypoxia.pathway)) + geom_point(size=0.5) +
  scale_color_gradientn(colours = c("blue", "blue", "white", "red", "red"), 
                        values = rescale(x = c(-1, 0, 0.5, 2.5, 6)),
                        name = 'PROGENy score') +
  ggtitle('Hypoxia pathway') +
  scale_x_continuous(limits= c(-6,-2)) + scale_y_continuous(limits= c(1,4)) + theme_minimal() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(size = 14),
        legend.key.size = unit(0.6, 'cm'),
        legend.text = element_text(size=8))

fig4h

```


## 4i

```{r panel_4i,fig.height=3,fig.width=4,eval=T}

## Plot PROGENy Hypoxia signature
fig4i <- ggplot(OV081_scRNA, aes(x=UMAP_1,y=UMAP_2,color=Hypoxia.pathway)) + geom_point(size=0.5) +
  scale_color_gradientn(colours = c("blue", "blue", "white", "red", "red"), 
                        values = rescale(x = c(-1, 0, 0.5, 2.5, 6)),
                        name = 'PROGENy score') +
  ggtitle('Hypoxia pathway') +
  scale_x_continuous(limits= c(-10.5,-6.5)) + scale_y_continuous(limits = c(10.5,14)) + theme_minimal() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(size = 14),
        legend.key.size = unit(0.6, 'cm'),
        legend.text = element_text(size=8))

fig4i

```



***

# Figure 5

```{r eval=T, echo=F}

# Import the master result file across all cells
resultfile = read.csv(paste0('../data/processed_data/final_master.tsv'), header=T, sep='\t', stringsAsFactors=F)
filtresultfile = resultfile[-which((resultfile$library_label == 'SA928-ALL') | (resultfile$group %in% c("treated TNBC PDX(SA1035)","treated TNBC PDX(SA535)","treated TNBC PDX(SA609)"))),]
filtresultfile = filtresultfile[-which(filtresultfile$clone_UMAP == 'nonmal'),]

# Import multiallele file
allvars = read.table(paste0('../data/processed_data/SA1047_genotype.tsv'),sep='\t',header=T,check.names=FALSE,stringsAsFactors=F)

# Import varmatrix file from Zenodo (https://zenodo.org/records/10535512)
varmatrix = read.table('../data/processed_data/varmatrix.tsv',header=T, sep='\t')

# Import the truncating variants matrix
truncvars = read.table('../data/processed_data/trunc_matrix.tsv',header=T, sep='\t')

# Import the expression correlation file
OV081_corr = read.table('../data/processed_data/OV-081_corr.txt',sep='\t',header=T, stringsAsFactors=F)

# Import the expression correlation file
combined_corr = read.table('../data/processed_data/combined_corr.txt',sep='\t',header=T, stringsAsFactors=F)

# Import the mitochondrial genes
genelist <- read.table("../data/ext/mtgenes.txt", header=T, quote = "", stringsAsFactors=F)
genelist <- genelist$KEGG_OXIDATIVE_PHOSPHORYLATION
MTgenes <- genelist[grep("MT-",genelist)]
nuMTgenes <- genelist[-grep("MT-",genelist)]

# Pathway enrichment analysis for clones
pathways.hallmark <- gmtPathways("../data/ext/h.all.v7.2.symbols.gmt")

```


## 5a

```{r panel_5a,fig.height=4,fig.width=4,eval=T}

## SA1047 m.302A vs. m.302A>AC
order1 = allvars$cell_id[order(allvars$`m.302A`)]
order2 = allvars$cell_id[order(allvars$`m.302A>AC`)]
order3 = allvars$cell_id[order(allvars$`m.302A>ACC`)]
order4 = allvars$cell_id[order(allvars$`m.302A>ACCC`)]
neworder = unique(c(order1,order2,order3,order4))
allvars$cell_id = factor(allvars$cell_id,level = rev(neworder))
allvars.m = reshape2::melt(allvars)
colnames(allvars.m) = c("cell_id","variant_id","heteroplasmy")

fig5a <- ggplot(data=allvars.m, aes(x=cell_id, y=heteroplasmy, fill=variant_id)) +
  geom_bar(position="fill", stat="identity") +
  scale_y_continuous(expand = c(0,0)) +
  xlab('Cells') + ylab('Heteroplasmy') + scale_fill_manual(values = customcolors$multiallele) + 
  guides(fill=guide_legend(title="Variant ID")) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text = element_text(size=6),
        axis.title = element_text(size=12),
        legend.key.size = unit(0.4,'cm'),
        legend.text = element_text(size=8),
        legend.title = element_blank(),
        legend.position = 'bottom',
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-5,0,0,-25))

fig5a

```


## 5b

```{r panel_5b,fig.height=2,fig.width=2,eval=T}

## Violin plot of the m.302A vs. m.302A>AC
filtvars = allvars[-which((allvars$`m.302A>ACC` > 0) | (allvars$`m.302A>ACCC` > 0)),]
filtvars$assignment = 'm.302A'
filtvars$heteroplasmy = filtvars$`m.302A`
filtvars$assignment[which(filtvars$`m.302A` < filtvars$`m.302A>AC`)] = 'm.302A>AC'
filtvars$heteroplasmy[which(filtvars$`m.302A` < filtvars$`m.302A>AC`)] = filtvars$`m.302A>AC`[which(filtvars$`m.302A` < filtvars$`m.302A>AC`)]
filtvars$copynumber = filtresultfile$copynumber[match(filtvars$cell_id,filtresultfile$cell_id)]
filtvars$ploidylab = filtresultfile$ploidylab[match(filtvars$cell_id,filtresultfile$cell_id)]
filtvars$clone_UMAP = filtresultfile$clone_UMAP[match(filtvars$cell_id,filtresultfile$cell_id)]

filtvarsummary = filtvars %>% group_by(assignment) %>% mutate(medianval = median(copynumber)) %>% select(assignment,medianval) %>% distinct()

fig5b <- ggplot(filtvars, aes(x=assignment,y=copynumber)) + geom_quasirandom(aes(color = assignment), width = 0.2, size = 0.2) +
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom = "crossbar", width = 0.3) +
  scale_color_manual(values = customcolors$multiallele) + theme_classic() + 
  stat_compare_means(label.x = 1.1, size = 2.5) + ylab('mtDNA copy number') +
  theme(legend.position = 'none',
        axis.title.x = element_blank(),
        axis.text = element_text(size=6),
        axis.title = element_text(size=8))
fig5b


```


## 5c

```{r panel_5c,fig.height=2.5,fig.width=2.5,eval=T}

## Heteroplasmy against m.302A
allvars$copynumber = resultfile$copynumber[match(allvars$cell_id,resultfile$cell_id)]

fig5c <- ggplot(data=allvars, aes(x=`m.302A`, y=copynumber)) +
  geom_point() + geom_smooth(method = 'lm') + stat_cor() +
  theme_classic() + xlab('Heteroplasmy') + ylab('mtDNA copy number')

fig5c

```


## 5d

```{r panel_5d,fig.height=2,fig.width=2,eval=T}

# Distinct library and variant
vargroups = truncvars %>% group_by(sample_id,variant_id) %>% mutate(start = strsplit(variant_id,':')[[1]][2]) %>% select(sample_id,variant_id,start) %>% distinct()
vargroups$start = as.numeric(vargroups$start)
vargroups_summary = vargroups %>% group_by(start) %>% mutate(n = n()) %>% select(start,n) %>% distinct()
vargroups_summary$start = as.numeric(vargroups_summary$start)

mitomap$xstart = mitomap$start
mitomap$xend = mitomap$end
mitomap$ystart = 0.255
mitomap$yend = 0.285
mitomap$ystart[which(mitomap$strand == -1)] = 0.225
mitomap$yend[which(mitomap$strand == -1)] = 0.255

y_offset <- 0.3

vargroups = merge(vargroups,vargroups_summary,by='start')
vargroups$n = vargroups$n/10 + y_offset
vargroups_summary$n = vargroups_summary$n/10 + 0.3

# plot
fig5d <- ggplot(mitomap) +
  geom_segment(data=vargroups,x=0.4,xend=0.4,y=1,yend=16569,color='grey',linetype=2,size=0.5) +
  geom_segment(data=vargroups,x=0.7,xend=0.7,y=1,yend=16569,color='grey',linetype=2,size=0.5) +
  geom_segment(data=vargroups,x=0.8*y_offset,xend=0.8*y_offset,y=1,yend=16569,color='black',size=0.5) +
  geom_segment(data=vargroups,x=0.9*y_offset,xend=0.9*y_offset,y=1,yend=16569,color='black',size=0.5) +
  
  geom_segment(data=vargroups,x=y_offset,
               aes(xend=n,y=start,yend=start),color='black',size=0.25) +
  geom_point(data=vargroups,
             aes(x=n,y=start),pch=21,color='black') +
  # Actual genes
  geom_rect(aes(ymin=xstart,ymax=xend,xmin=ystart,xmax=yend,fill=gene),color='black',size=0.5) +
  
  # Customization
  scale_x_continuous(limits=c(0,max(vargroups$n))) +
  scale_y_continuous(position='left') +
  coord_polar(theta="y") + theme_classic() +
  theme(
    axis.line.x=element_blank(), axis.ticks.x=element_blank(), axis.text.x=element_blank(),
    axis.line.y=element_blank(),axis.ticks.y=element_blank(), axis.text.y=element_blank(),
    legend.position = 'none') +
  geom_text_repel(data=vargroups_summary,color='black',aes(x=n,y=start,label=start),size=2) +
  labs(x=NULL,y=NULL,title=NULL)

fig5d


```



## 5e

```{r panel_5e,fig.height=3,fig.width=4,eval=T}

# Varmatrix
filtvarmatrix = varmatrix
filtvarmatrix = filtvarmatrix[-which(filtvarmatrix$heteroplasmy > 0.95),]
filtvarmatrix = filtvarmatrix[-which(filtvarmatrix$heteroplasmy < 0.05),]
if (length(which(is.na(filtvarmatrix$mtDNAcopynumber))) > 0){
  filtvarmatrix = filtvarmatrix[-which(is.na(filtvarmatrix$mtDNAcopynumber)),]
}

# Filtering by threshold
readthreshold = 10
filtvarmatrix = filtvarmatrix[which(filtvarmatrix$totalread >= readthreshold),]

# Keep only the malignant cells
if (length(which(filtvarmatrix$clone_UMAP == 'nonmal')) > 0){
  filtvarmatrix = filtvarmatrix[-which(filtvarmatrix$clone_UMAP == 'nonmal'),]
}

# Filter based on heteroplasmy range
filtvarmatrix = filtvarmatrix %>% group_by(library_id,variant_id) %>% dplyr::mutate(hetrange = max(heteroplasmy) - min(heteroplasmy)) %>% dplyr::filter(hetrange >= 0.15) %>% select(library_id,variant_id,cell_id,ploidylab,heteroplasmy,altread,totalread,mtDNAcopynumber,type,group,datatype,sample_id,clone_UMAP,hetrange) %>% distinct()

addtocombinedres <- function(eachtype,eachvartype,eachvar,filtcurrentres,currentsample,whichtype){
  if(whichtype == 'group_id'){
    subsetcurrentres = filtcurrentres[grep(currentsample,filtcurrentres$group),]
    
    # Determine the weights based on 1/n
    subsetcurrentres = subsetcurrentres %>% dplyr::group_by(ploidylab,library_id) %>% dplyr::mutate(n=n(),weight = n/nrow(subsetcurrentres))
    
    if (nrow(subsetcurrentres) >= 10){
      # Determine the concordance value, accounting for the ploidy (strata(ploidylab)), library ID (library_id), clone ID (clone_UMAP)
      concres = survival::concordance(mtDNAcopynumber ~ heteroplasmy + strata(ploidylab) + strata(library_id), data=subsetcurrentres, weights = weight)
      
      # p-value calculation based on z-score
      zscore = (concres$concordance - 0.5)/sqrt(concres$var)
      p_value = 2*pnorm(-abs(zscore))
      
      # Store result
      tempres = data.frame(eachtype,currentsample,eachvartype,eachvar,concres$concordance,concres$var,p_value)
      colnames(tempres) = c('datatype','sample_id','vartype','variant_id','concordance','var','pval')
      return(tempres)
    }
  }
  else if(whichtype == 'sample_id'){
    tottempres = data.frame(t(rep(1,7)))
    colnames(tottempres) = c('datatype','sample_id','vartype','variant_id','concordance','var','pval')
    for (eachsample in currentsample){
      subsetcurrentres = filtcurrentres[which(filtcurrentres$sample_id == eachsample),]
      
      # Determine the weights based on 1/n
      subsetcurrentres = subsetcurrentres %>% dplyr::group_by(ploidylab,library_id) %>% dplyr::mutate(n=n(),weight = n/nrow(subsetcurrentres))
      
      if (nrow(subsetcurrentres) >= 10){
        # Determine the concordance value, accounting for the ploidy (strata(ploidylab)), library ID (library_id), clone ID (clone_UMAP)
        concres = survival::concordance(mtDNAcopynumber ~ heteroplasmy + strata(ploidylab) + strata(library_id), data=subsetcurrentres, weights = weight)
        
        # p-value calculation based on z-score
        zscore = (concres$concordance - 0.5)/sqrt(concres$var)
        p_value = 2*pnorm(-abs(zscore))
        
        # Store result
        tempres = data.frame(eachtype,eachsample,eachvartype,eachvar,concres$concordance,concres$var,p_value)
        colnames(tempres) = c('datatype','sample_id','vartype','variant_id','concordance','var','pval')
        tottempres = rbind(tottempres, tempres)
      }
    }
    tottempres = tottempres[-1,]
    return(tottempres)
  }
}

# Initialize the result dataframe
combinedres = data.frame(t(rep(1,7)))
colnames(combinedres) = c('datatype','sample_id','vartype','variant_id','concordance','var','pval')

for (eachtype in c('cell_lines','tumors')){
  for (eachvartype in c('silent','truncating','nontruncating')){
    filtres = filtvarmatrix[which((filtvarmatrix$datatype == eachtype) & (filtvarmatrix$type == eachvartype)),]
    
    for (eachvar in unique(filtres$variant_id)){
      currsample = unique(filtres$sample_id[which(filtres$variant_id == eachvar)])
      if(length(currsample) > 1){ # more than 1 sample
        if (length(currsample[which(currsample %in% hTERT_group)]) > 0){ # hTERT samples
          if (length(currsample[-which(currsample %in% hTERT_group)]) > 0){ # other than hTERT samples
            # iterate through the other samples
            for (othersample in currsample[-which(currsample %in% hTERT_group)]){
              filtcurrentres = filtres[which((filtres$variant_id == eachvar) & (filtres$sample_id == othersample)),]
              tempres = addtocombinedres(eachtype,eachvartype,eachvar,filtcurrentres,othersample,whichtype = 'sample_id')
              combinedres = rbind(combinedres,tempres)
            }
            # hTERT samples
            filtcurrentres = filtres[which((filtres$variant_id == eachvar) & (filtres$sample_id %in% hTERT_group)),]
            tempres = addtocombinedres(eachtype,eachvartype,eachvar,filtcurrentres,'184hTERT',whichtype = 'group_id')
            combinedres = rbind(combinedres,tempres)
          }else{
            # hTERT samples only but multiple samples
            filtcurrentres = filtres[which((filtres$variant_id == eachvar) & (filtres$sample_id %in% hTERT_group)),]
            tempres = addtocombinedres(eachtype,eachvartype,eachvar,filtcurrentres,'184hTERT',whichtype = 'group_id')
            combinedres = rbind(combinedres,tempres)
          }
        }else{
          # iterate through the samples
          for (onesample in currsample){
            filtcurrentres = filtres[which((filtres$variant_id == eachvar) & (filtres$sample_id == onesample)),]
            tempres = addtocombinedres(eachtype,eachvartype,eachvar,filtcurrentres,onesample,whichtype = 'sample_id')
            combinedres = rbind(combinedres,tempres)
          }
        }
      }else{ # Single sample variant
        filtcurrentres = filtres[which(filtres$variant_id == eachvar),]
        currentsample = unique(filtcurrentres$sample_id)
        tempres = addtocombinedres(eachtype,eachvartype,eachvar,filtcurrentres,currentsample,whichtype = 'sample_id')
        combinedres = rbind(combinedres,tempres)
      }
    }
  }
}
combinedres = combinedres[-1,]

# Change the labels
ref = apply(data.frame(combinedres$variant_id),1,function(x) strsplit(x,':')[[1]][1])
pos = as.numeric(apply(data.frame(combinedres$variant_id),1,function(x) strsplit(x,':')[[1]][2]))
alt = apply(data.frame(combinedres$variant_id),1,function(x) strsplit(x,':')[[1]][3])
combinedres$variant_id = paste0('m.',pos,ref,'>',alt)

# Adjusted p-values
combinedres$padj = p.adjust(combinedres$pval,method='BH')
combinedres$p10 = -log10(combinedres$padj)
combinedres$signif = combinedres$padj <= 0.05
combinedres$label = ''
combinedres$label[which((combinedres$padj < 0.15) |
                          (combinedres$concordance > 0.55) |
                          (combinedres$concordance < 0.4))] = combinedres$variant_id[which((combinedres$padj < 0.15) |
                                                                                             (combinedres$concordance > 0.55) |
                                                                                             (combinedres$concordance < 0.4))]

# Plot the volcano plot
fig5e <- ggplot(combinedres, aes(x = concordance, y = p10, label = label, color = vartype, alpha = signif)) +
  geom_vline(xintercept = 0.5, color = 'grey') +
  geom_hline(yintercept = 0, color = 'grey') +
  geom_point(aes(shape = datatype), size = 1.4) +
  scale_alpha_manual(values=c(0.4,1)) +
  scale_shape_manual(values=c(2,16)) +
  scale_color_manual(values = c("#0186BA",'#CE2B19',"#D38842")) +
  geom_text_repel(size = 2) +
  scale_x_continuous(limits = c(0.3,0.8), expand=c(0,0)) +
  scale_y_continuous(limits = c(0,2.6), expand=c(0,0)) +
  geom_hline(yintercept = -log10(0.05), 
             linetype = 'dashed', color = 'grey') +
  guides(alpha=guide_legend(title = 'Significant'),color=guide_legend(title = 'Variant type'),shape=guide_legend(title = 'Data type')) +
  xlab('Concordance') +
  ylab('-log10(Q-value)') +
  theme_classic() +
  theme(axis.text = element_text(size=6),
        axis.title = element_text(size=8),
        legend.key.size = unit(0.4,'cm'),
        legend.title = element_text(size=6),
        legend.text = element_text(size=6),
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(0,0,0,-5))
fig5e

```

## 5f and 5g from IGV

## 5h

```{r panel_5h,fig.height=2,fig.width=2.5,eval=T}

# varmatrix
filtvarmatrix = varmatrix
filtvarmatrix = filtvarmatrix[-which(filtvarmatrix$heteroplasmy > 0.95),]
filtvarmatrix = filtvarmatrix[-which(filtvarmatrix$heteroplasmy < 0.05),]
if (length(which(is.na(filtvarmatrix$mtDNAcopynumber))) > 0){
  filtvarmatrix = filtvarmatrix[-which(is.na(filtvarmatrix$mtDNAcopynumber)),]
}

if (length(which(filtvarmatrix$clone_UMAP == 'nonmal')) > 0){
  filtvarmatrix = filtvarmatrix[-which(filtvarmatrix$clone_UMAP == 'nonmal'),]
}

# Filter based on heteroplasmy range
filtvarmatrix = filtvarmatrix %>% group_by(library_id,variant_id) %>% dplyr::mutate(hetrange = max(heteroplasmy) - min(heteroplasmy)) %>% dplyr::filter(hetrange >= 0.15) %>% select(library_id,variant_id,cell_id,ploidylab,heteroplasmy,altread,totalread,mtDNAcopynumber,type,group,datatype,sample_id,clone_UMAP,hetrange) %>% distinct()

filtvarmatrix = filtvarmatrix[which((filtvarmatrix$sample_id == "SPECTRUM-OV-081") & (filtvarmatrix$variant_id == "G:6708:A")),]

fig5h <- ggplot(filtvarmatrix,aes(x=heteroplasmy,y=mtDNAcopynumber)) + geom_point(size=0.6) + theme_classic() +
  geom_smooth(method = 'lm') + stat_cor() +
  xlab('Heteroplasmy') + ylab('mtDNA copy number') +
  theme(strip.text.x = element_text(size=10),
        strip.background = element_blank())

fig5h


```


## 5i

```{r panel_5i,fig.height=2.5,fig.width=2.5,eval=T}

# Heteroplasmy correlation
fig5i <- ggplot(subset(OV081_corr,gene_id == 'MT-ND3'),aes(x=Heteroplasmy,y=gene_exp)) + geom_point() +
  theme_classic() + geom_smooth(method='lm') + stat_cor() + 
  ylab('MT-ND3 Expression')

fig5i


```



## 5j

```{r panel_5j,fig.height=2.5,fig.width=2.5,eval=T}

filtcombined_corr = combined_corr[which((combined_corr$sample_id == 'SPECTRUM-OV-081') & (combined_corr$variant_id == 'G:6708:A')),]
filtcombined_corr = filtcombined_corr[order(filtcombined_corr$coeff,decreasing = T),]
if (length(which(is.na(filtcombined_corr$coeff))) > 0){
  filtcombined_corr = filtcombined_corr[-which(is.na(filtcombined_corr$coeff)),]
}
filtcombined_corr$Index = 1:nrow(filtcombined_corr)
filtcombined_corr$genome = 'nuDNA'
filtcombined_corr$genome[grep('MT-',filtcombined_corr$gene_id)] = 'mtDNA'
MTres = filtcombined_corr[which(filtcombined_corr$genome == 'mtDNA'),]
filtcombined_corr = filtcombined_corr[-which(filtcombined_corr$genome == 'mtDNA'),]
filtcombined_corr$label = ''
MTres$label = MTres$gene_id

fig5j <- ggplot(filtcombined_corr, aes(x=Index,y=coeff,color=genome)) + geom_point(size=0.2) + geom_hline(yintercept = 0, linetype = 'dashed') +
  scale_color_manual(values=c('nuDNA'='black','mtDNA'='red')) + theme_classic() +
  geom_point(data=MTres,aes(x=Index,y=coeff,color=genome),inherit.aes = F) +
  xlab('Rank sorted genes') + ylab('Correlation coefficient with heteroplasmy') +
  theme(legend.position=c(0.2,0.2),
        axis.title = element_text(size=8),
        axis.text = element_text(size=8),
        legend.key.size = unit(0.4,'cm'),
        legend.title = element_blank(),
        legend.text = element_text(size=6))
fig5j

```



## 5k

```{r panel_5k,fig.height=3,fig.width=2.5,eval=T}

combinedres = data.frame(t(rep(1,9)))
colnames(combinedres) = c("sample","pathway","pval","padj","log2err","ES","NES","size","stat")

resummary = combined_corr %>% group_by(sample_id,variant_id) %>% select(sample_id,variant_id) %>% distinct()

filtres = combined_corr[which((combined_corr$sample_id == 'SPECTRUM-OV-081') & (combined_corr$variant_id == 'G:6708:A')),]

checkres = filtres %>% group_by(gene_id) %>% mutate(medcoeff = median(coeff,na.rm=T)) %>% select(gene_id,medcoeff) %>% distinct()
checkres = checkres[order(checkres$medcoeff,decreasing = T),]
if (length(which(is.na(checkres$medcoeff))) > 0){
  checkres = checkres[-which(is.na(checkres$medcoeff)),]
}
checkres$Index = 1:nrow(checkres)
checkres$genome = 'Nuclear'
checkres$genome[grep('MT-',checkres$gene_id)] = 'MT'
MTcheckres = checkres[which(checkres$genome == 'MT'),]
checkres = checkres[-which(checkres$genome == 'MT'),]
checkres$label = ''
MTcheckres$label = MTcheckres$gene_id

rownames(filtres) = filtres$gene_id
filtres$MTgenes = "None"
filtres[intersect(filtres$gene_id, MTgenes),"MTgenes"] = "mtDNA-encoded"
filtres[intersect(filtres$gene_id, nuMTgenes),"MTgenes"] = "nuDNA-encoded"
rank = filtres[order(filtres$coeff, decreasing=T),]
if (length(which(is.na(rank$coeff))) > 0){
  rank = rank[-which(is.na(rank$coeff)),]
}
ranks <- rank$coeff
names(ranks) <- rank$gene_id

fgseaRes <- fgseaMultilevel(pathways=pathways.hallmark, stats=ranks)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy$pathway = gsub('HALLMARK_','',fgseaResTidy$pathway)
fgseaResTidy$pathway = gsub('_',' ',fgseaResTidy$pathway)
fgseaResTidy$Significant = fgseaResTidy$padj<=0.05
fgseaResTidy$Significant = factor(fgseaResTidy$Significant, levels=c("TRUE","FALSE"))

fgseaResTidy$stat = -log10(fgseaResTidy$padj)*sign(fgseaResTidy$NES)
fgseaResTidy$sample = 'SPECTRUM-OV-081'
fgseaResTidy = fgseaResTidy[,c("sample","pathway","pval","padj","log2err","ES","NES","size","stat")]

combinedres = fgseaResTidy
combinedres$pathway = gsub('HALLMARK_','',combinedres$pathway)
combinedres$pathway = gsub('_',' ',combinedres$pathway)

summarize_signature <- function(m) {
  mid <- median(m$stat)
  n <- sum(abs(m$stat) > 2) #/ nrow(m)
  m <- m[order(stat,decreasing=F),]
  m$n <- n
  m$mid <- mid
  m$pos <- ((1:nrow(m))/nrow(m))
  m
}

combinedres <- adt(combinedres)
combinedres <- combinedres[,summarize_signature(.SD), by=pathway]
combinedres <- combinedres[order(mid,pos,decreasing=F),]
combinedres <- combinedres[which((combinedres$stat <= -2) | (combinedres$stat >= 2)),]
combinedres$pathway <- factor(combinedres$pathway, levels=unique(combinedres$pathway))

fig5k <- ggplot(combinedres, aes(x=stat,y=pathway)) +
  scale_x_continuous(limits=c(-10,20)) + 
  geom_bar(stat='identity',fill='#4675b4') +
  theme_classic() +
  theme(axis.text.y = element_text(size=6),
        axis.title = element_text(size=8),
        legend.key.size = unit(0.3,'cm'),
        legend.key.height = unit(0.3,'cm'),
        legend.key.width = unit(0.3,'cm'),
        legend.title = element_text(size=5),
        legend.text = element_text(size=5)) + 
  labs(x=expression('-log10(Q-value)*sign(NES)'),y=NULL)

fig5k

```




